<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>Inspect S1 sequencing and qPCR results</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-09-28"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-09-28"/>
  <meta name="article:author" content="Michael R. McLaren"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Inspect S1 sequencing and qPCR results"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Inspect S1 sequencing and qPCR results"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","categories","date","output"]}},"value":[{"type":"character","attributes":{},"value":["Inspect S1 sequencing and qPCR results"]},{"type":"NULL"},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Michael R. McLaren"]}]}]},{"type":"character","attributes":{},"value":["2021 experiment"]},{"type":"character","attributes":{},"value":["2021-09-28"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["inspect-s1-sequencing-results_files/anchor-4.2.2/anchor.min.js","inspect-s1-sequencing-results_files/bowser-1.9.3/bowser.min.js","inspect-s1-sequencing-results_files/distill-2.2.21/template.v2.js","inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-15-1.png","inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-18-1.png","inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-20-1.png","inspect-s1-sequencing-results_files/header-attrs-2.9/header-attrs.js","inspect-s1-sequencing-results_files/jquery-1.11.3/jquery.min.js","inspect-s1-sequencing-results_files/popper-2.6.0/popper.min.js","inspect-s1-sequencing-results_files/tippy-6.2.7/tippy-bundle.umd.min.js","inspect-s1-sequencing-results_files/tippy-6.2.7/tippy-light-border.css","inspect-s1-sequencing-results_files/tippy-6.2.7/tippy.css","inspect-s1-sequencing-results_files/tippy-6.2.7/tippy.umd.min.js","inspect-s1-sequencing-results_files/webcomponents-2.0.0/webcomponents.js"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="inspect-s1-sequencing-results_files/header-attrs-2.9/header-attrs.js"></script>
  <script src="inspect-s1-sequencing-results_files/jquery-1.11.3/jquery.min.js"></script>
  <script src="inspect-s1-sequencing-results_files/popper-2.6.0/popper.min.js"></script>
  <link href="inspect-s1-sequencing-results_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="inspect-s1-sequencing-results_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="inspect-s1-sequencing-results_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="inspect-s1-sequencing-results_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="inspect-s1-sequencing-results_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="inspect-s1-sequencing-results_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="inspect-s1-sequencing-results_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Inspect S1 sequencing and qPCR results","authors":[{"author":"Michael R. McLaren","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2021-09-28T00:00:00.000-04:00","citationText":"McLaren, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Inspect S1 sequencing and qPCR results</h1>
<!--radix_placeholder_categories-->
<div class="dt-tags">
<div class="dt=tag">2021 experiment</div>
</div>
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Michael R. McLaren  
  
<br/>2021-09-28
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#load-data">load data</a>
<ul>
<li><a href="#sample-data">Sample data</a></li>
<li><a href="#import-the-picogreen-dna-concentration-measurements">Import the PicoGreen DNA concentration measurements</a></li>
<li><a href="#qpcr-results">qPCR results</a></li>
<li><a href="#reads">Reads</a></li>
<li><a href="#combine">combine</a></li>
</ul></li>
<li><a href="#analysis">Analysis</a>
<ul>
<li><a href="#reads-1">Reads</a></li>
<li><a href="#session-info">Session info</a></li>
</ul></li>
</ul>
</nav>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; ── Attaching packages ─────────────────────────────────────────── tidyverse 1.3.1 ──
#&gt; ✔ ggplot2 3.3.5     ✔ purrr   0.3.4
#&gt; ✔ tibble  3.1.3     ✔ dplyr   1.0.7
#&gt; ✔ tidyr   1.1.3     ✔ stringr 1.4.0
#&gt; ✔ readr   2.0.0     ✔ forcats 0.5.1
#&gt; ── Conflicts ────────────────────────────────────────────── tidyverse_conflicts() ──
#&gt; ✖ dplyr::filter() masks stats::filter()
#&gt; ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; here() starts at /home/michael/research/vivo-vitro</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://fs.r-lib.org'>fs</a></span><span class='op'>)</span>
<span class='co'># library(tabulizer)</span>

<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; 
#&gt; Attaching package: &#39;patchwork&#39;
#&gt; The following object is masked from &#39;package:cowplot&#39;:
#&gt; 
#&gt;     align_plots</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>theme_set</span><span class='op'>(</span><span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'>import</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/import/man/importfunctions.html'>from</a></span><span class='op'>(</span><span class='va'>colorblindr</span>, <span class='va'>scale_color_OkabeIto</span>, <span class='va'>scale_fill_OkabeIto</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Note, to successfully install tabulizer I needed to switch from OpenJDK 8 to OpenJDK 11.</p>
<h1 id="load-data">load data</h1>
<h2 id="sample-data">Sample data</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>'data/2021/'</span>, <span class='st'>'sample-data'</span>, <span class='st'>'dna-sample-data.csv'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='va'>read_csv</span>
</code></pre>
</div>
<pre><code>#&gt; Rows: 94 Columns: 14
#&gt; ── Column specification ────────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (8): row, well, dna_sample_id, specimen_name, mixture, specime...
#&gt; dbl (5): column, dilution_power, dilution_factor, aliquot_number, ...
#&gt; lgl (1): feces_added
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<h2 id="import-the-picogreen-dna-concentration-measurements">Import the PicoGreen DNA concentration measurements</h2>
<p>These were provided as tables in pdf files. Let’s try to use the R package tabulizer to read these tables into R.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pcg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>'data/2021/s1'</span>, 
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'NCSUMclaren_Plate_0002 PicoGreen QC.pdf'</span>, 
    <span class='st'>'NCSUMclaren_Plate_0003 PicoGreen QC.pdf'</span>
  <span class='op'>)</span>
<span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>map</span><span class='op'>(</span><span class='fu'>tabulizer</span><span class='fu'>::</span><span class='va'><a href='https://docs.ropensci.org/tabulizer/reference/extract_tables.html'>extract_tables</a></span>, output <span class='op'>=</span> <span class='st'>'data.frame'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>map</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>map</span><span class='op'>(</span><span class='va'>as_tibble</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>bind_rows</span><span class='op'>(</span>.id <span class='op'>=</span> <span class='st'>'plate'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>janitor</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/janitor/man/clean_names.html'>clean_names</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pcg</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>sample_id</span> <span class='op'>==</span> <span class='st'>'Empty'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>count</span><span class='op'>(</span><span class='va'>concentration_ng_ul</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 1 × 2
#&gt;   concentration_ng_ul     n
#&gt;   &lt;chr&gt;               &lt;int&gt;
#&gt; 1 Undetected             98</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pcg</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>sample_id</span> <span class='op'>==</span> <span class='st'>''</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 4 × 5
#&gt;   plate sample_id well  concentration_ng_ul comments
#&gt;   &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;               &lt;chr&gt;   
#&gt; 1 1     &quot;&quot;        &quot;&quot;    &quot;&quot;                  &quot;&quot;      
#&gt; 2 1     &quot;&quot;        &quot;&quot;    &quot;&quot;                  &quot;&quot;      
#&gt; 3 2     &quot;&quot;        &quot;&quot;    &quot;&quot;                  &quot;&quot;      
#&gt; 4 2     &quot;&quot;        &quot;&quot;    &quot;&quot;                  &quot;&quot;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pcg1</span> <span class='op'>&lt;-</span> <span class='va'>pcg</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='op'>!</span><span class='op'>(</span><span class='va'>sample_id</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'Empty'</span>, <span class='st'>''</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>rename</span><span class='op'>(</span>
    dna_sample_id <span class='op'>=</span> <span class='va'>sample_id</span>,
    dna_conc <span class='op'>=</span> <span class='va'>concentration_ng_ul</span>, 
    well3 <span class='op'>=</span> <span class='va'>well</span>,
  <span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span>
    dna_conc <span class='op'>=</span> <span class='fu'>case_when</span><span class='op'>(</span>
      <span class='va'>dna_conc</span> <span class='op'>==</span> <span class='st'>'Undetected'</span> <span class='op'>~</span> <span class='st'>'0'</span>,
      <span class='cn'>TRUE</span> <span class='op'>~</span> <span class='va'>dna_conc</span>
    <span class='op'>)</span>,
    <span class='fu'>across</span><span class='op'>(</span><span class='va'>dna_conc</span>, <span class='va'>as.numeric</span><span class='op'>)</span>,
    <span class='fu'>across</span><span class='op'>(</span><span class='va'>dna_sample_id</span>, <span class='va'>str_replace_all</span>, <span class='st'>'_'</span>, <span class='st'>'-'</span><span class='op'>)</span>
  <span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="qpcr-results">qPCR results</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qpcr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>'data/2021/s1'</span>, 
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'NCSUMcLaren_Plate_0002_AbsQuant.csv'</span>, 
    <span class='st'>'NCSUMcLaren_Plate_0003_AbsQuant.csv'</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>map</span><span class='op'>(</span><span class='va'>read_csv</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>bind_rows</span><span class='op'>(</span>.id <span class='op'>=</span> <span class='st'>'plate'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>janitor</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/janitor/man/clean_names.html'>clean_names</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Rows: 96 Columns: 5
#&gt; ── Column specification ────────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (2): SampleWellThreeCharacter, SampleID
#&gt; dbl (3): fkPLATE_ID, cCopies, cCV
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#&gt; Rows: 96 Columns: 5
#&gt; ── Column specification ────────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (2): SampleWellThreeCharacter, SampleID
#&gt; dbl (3): fkPLATE_ID, cCopies, cCV
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qpcr</span> <span class='op'>%&gt;%</span>
  <span class='fu'>count</span><span class='op'>(</span><span class='va'>plate</span>, <span class='va'>fk_plate_id</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 2 × 3
#&gt;   plate fk_plate_id     n
#&gt;   &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
#&gt; 1 1               1    96
#&gt; 2 2               1    96</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qpcr</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>sample_id</span><span class='op'>)</span> <span class='op'>&amp;</span> <span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>c_copies</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 3 × 6
#&gt;   plate fk_plate_id sample_well_three_charac… sample_id c_copies  c_cv
#&gt;   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;                     &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;
#&gt; 1 1               1 C10                       F2_D2_1         NA    NA
#&gt; 2 1               1 D08                       F3_D2_1         NA    NA
#&gt; 3 2               1 E02                       F1_D2_2         NA    NA</code></pre>
</div>
<p>These three samples have no qPCR measurements; these are all high-dilution fecal samples. My guess is that the 16s CN was below detection limit, or otherwise the qPCR failed due to low CN. Confirm with Eric, and ask if he has any info on what the detection limit is.</p>
<p>Should ask exactly what the ‘c_copies’ and ‘c_cv’ files are; cv is presumably coefficient of variation - is this the standard error in ‘c_copies’ divided by ‘c_copies’? I think that the estimated copies are per a specific volume - should check which, which I think came up in the email thread.</p>
<p>todo: do more checks, such as that the right samples are in the right plates</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qpcr1</span> <span class='op'>&lt;-</span> <span class='va'>qpcr</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>sample_id</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>rename</span><span class='op'>(</span>
    dna_sample_id <span class='op'>=</span> <span class='va'>sample_id</span>,
    well3 <span class='op'>=</span> <span class='va'>sample_well_three_character</span>,
    qpcr_copies <span class='op'>=</span> <span class='va'>c_copies</span>, 
    qpcr_cv <span class='op'>=</span> <span class='va'>c_cv</span>
  <span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span>
    <span class='fu'>across</span><span class='op'>(</span><span class='va'>dna_sample_id</span>, <span class='va'>str_replace_all</span>, <span class='st'>'_'</span>, <span class='st'>'-'</span><span class='op'>)</span>,
  <span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>replace_na</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>qpcr_copies <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>select</span><span class='op'>(</span><span class='op'>-</span><span class='va'>fk_plate_id</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="reads">Reads</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>overview</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>'data/2021/s1'</span>, <span class='st'>'sample_overview.txt'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='va'>read_tsv</span> <span class='op'>%&gt;%</span>
  <span class='fu'>janitor</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/janitor/man/clean_names.html'>clean_names</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Rows: 89 Columns: 7
#&gt; ── Column specification ────────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;\t&quot;
#&gt; chr (1): Sample_ID
#&gt; dbl (6): Read_Depth, Reads_Mapping, Reads_Genus_Level, Reads_Speci...
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<p>note, only 89 samples; some were not processed in their pipeline due to insufficient reads / quality.</p>
<p>might be better to instead calculate the reads w/ a tool; I think I have some code to do this with the shortread package somewhere.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>overview1</span> <span class='op'>&lt;-</span> <span class='va'>overview</span> <span class='op'>%&gt;%</span>
  <span class='fu'>select</span><span class='op'>(</span>
    dna_sample_id <span class='op'>=</span> <span class='va'>sample_id</span>,
    <span class='va'>read_depth</span>
  <span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span>
    <span class='fu'>across</span><span class='op'>(</span><span class='va'>dna_sample_id</span>, <span class='va'>str_replace_all</span>, <span class='st'>'\\.'</span>, <span class='st'>'-'</span><span class='op'>)</span>,
  <span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="combine">combine</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>&lt;-</span> <span class='va'>sam</span> <span class='op'>%&gt;%</span>
  <span class='fu'>left_join</span><span class='op'>(</span><span class='va'>pcg1</span>, by <span class='op'>=</span> <span class='st'>'dna_sample_id'</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>left_join</span><span class='op'>(</span><span class='va'>qpcr1</span>, by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'plate'</span>, <span class='st'>'well3'</span>, <span class='st'>'dna_sample_id'</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>left_join</span><span class='op'>(</span><span class='va'>overview1</span>, by <span class='op'>=</span> <span class='st'>'dna_sample_id'</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h1 id="analysis">Analysis</h1>
<p>Compare qPCR and PicoGreen measurements</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>%&gt;%</span> <span class='fu'>count</span><span class='op'>(</span><span class='va'>dilution_power</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 3 × 2
#&gt;   dilution_power     n
#&gt;            &lt;dbl&gt; &lt;int&gt;
#&gt; 1              0    34
#&gt; 2              1    30
#&gt; 3              2    30</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>%&gt;%</span> <span class='fu'>count</span><span class='op'>(</span><span class='va'>specimen_type</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 4 × 2
#&gt;   specimen_type     n
#&gt;   &lt;chr&gt;         &lt;int&gt;
#&gt; 1 Fecal            30
#&gt; 2 Mock in feces    30
#&gt; 3 Mock only        30
#&gt; 4 T. mobilis        4</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>%&gt;%</span>
  <span class='fu'>ggplot</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span><span class='va'>dna_conc</span> <span class='op'>+</span> <span class='fl'>3e-2</span>, <span class='va'>qpcr_copies</span> <span class='op'>+</span> <span class='fl'>1e5</span>, color <span class='op'>=</span> <span class='va'>specimen_type</span>, 
      label <span class='op'>=</span> <span class='va'>dilution_power</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_text</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_x_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_y_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_color_OkabeIto</span><span class='op'>(</span>order <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>8</span><span class='op'>)</span>, use_black <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_minimal_grid.html'>theme_minimal_grid</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>coord_fixed</span><span class='op'>(</span>clip <span class='op'>=</span> <span class='st'>'off'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-15-1.png" width="672" /></p>
</div>
<p>Good concordance. Note that there are three overplotted samples where no DNA was measured by either method.</p>
<p>Compare qPCR values versus dilution for each specimen. Note - we have replicate aliquots in all cases, so we’ll want to first average over those.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://mjskay.github.io/ggdist/'>ggdist</a></span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>I’ll just use the median; a better job might be to do a weighted geometric mean where we use the CV for the weights (though we have structure because of the two extraction protocols; it might be useful to not combine these actually…)</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='va'>sam1</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span><span class='fu'>across</span><span class='op'>(</span><span class='va'>qpcr_copies</span>, <span class='va'>pmax</span>, <span class='fl'>1e5</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>specimen_name</span>, <span class='va'>specimen_base_id</span>, <span class='va'>specimen_type</span>, <span class='va'>dilution_factor</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='http://mjskay.github.io/ggdist/reference/point_interval.html'>point_interval</a></span><span class='op'>(</span><span class='va'>qpcr_copies</span>, .width <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>%&gt;%</span> 
  <span class='co'># add horizontal jitter</span>
  <span class='fu'>mutate</span><span class='op'>(</span><span class='fu'>across</span><span class='op'>(</span><span class='va'>dilution_factor</span>, <span class='op'>~</span> <span class='va'>.</span> <span class='op'>*</span> <span class='fl'>10</span><span class='op'>^</span><span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span><span class='op'>(</span><span class='fu'>n</span><span class='op'>(</span><span class='op'>)</span>, sd <span class='op'>=</span> <span class='fl'>0.02</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>ggplot</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span><span class='va'>dilution_factor</span>, <span class='va'>qpcr_copies</span>, color <span class='op'>=</span> <span class='va'>specimen_type</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_line</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>group <span class='op'>=</span> <span class='va'>specimen_base_id</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='http://mjskay.github.io/ggdist/reference/geom_pointinterval.html'>geom_pointinterval</a></span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>ymin <span class='op'>=</span> <span class='va'>.lower</span>, ymax <span class='op'>=</span> <span class='va'>.upper</span><span class='op'>)</span>, 
    point_size <span class='op'>=</span> <span class='fl'>2</span>, interval_size <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_x_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_y_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_color_OkabeIto</span><span class='op'>(</span>order <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>8</span><span class='op'>)</span>, use_black <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_minimal_grid.html'>theme_minimal_grid</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>coord_fixed</span><span class='op'>(</span>clip <span class='op'>=</span> <span class='st'>'off'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-18-1.png" width="672" /></p>
</div>
<p>These results indicate that the DNA yield overall was quite linear in input. Should explore this further, perhaps breaking down by protocol. But looks very promising for the value of qPCR for relative changes in input concentration when the composition (relative abundances) is consistent.</p>
<p>These results also suggest that adding feces had little impact on bacterial extraction yield.</p>
<p>These results also show that we succeeded in our aim of getting the concentration s of the mock and fecal types to overlap; the lowest dilution of the mocks gets below the undiluted fecal samples.</p>
<p>todo - investigate the outlier</p>
<p>Note, these error bars are misleadingly long b/c of how I grouped by extraction protocol.</p>
<p>todo: in this case we should use a geometric mean -</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='va'>sam1</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span><span class='fu'>across</span><span class='op'>(</span><span class='va'>qpcr_copies</span>, <span class='va'>pmax</span>, <span class='fl'>1e5</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>specimen_name</span>, <span class='va'>specimen_base_id</span>, <span class='va'>specimen_type</span>, <span class='va'>dilution_factor</span>,
    <span class='va'>extraction_protocol</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='http://mjskay.github.io/ggdist/reference/point_interval.html'>point_interval</a></span><span class='op'>(</span><span class='va'>qpcr_copies</span>, .width <span class='op'>=</span> <span class='fl'>1</span>, .point <span class='op'>=</span> <span class='fu'>metacal</span><span class='fu'>::</span><span class='va'><a href='https://mikemc.github.io/metacal/reference/gm_mean.html'>gm_mean</a></span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>%&gt;%</span> 
  <span class='co'># add horizontal jitter</span>
  <span class='co'># mutate(across(dilution_factor, ~ . * 10^rnorm(n(), sd = 0.02))) %&gt;%</span>
  <span class='fu'>ggplot</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span><span class='va'>dilution_factor</span>, <span class='va'>qpcr_copies</span>, color <span class='op'>=</span> <span class='va'>specimen_type</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>facet_wrap</span><span class='op'>(</span><span class='op'>~</span><span class='va'>extraction_protocol</span>, labeller <span class='op'>=</span> <span class='st'>'label_both'</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_line</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>group <span class='op'>=</span> <span class='va'>specimen_base_id</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_point</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='co'># geom_pointinterval(aes(ymin = .lower, ymax = .upper), </span>
  <span class='co'>#   point_size = 2, interval_size = 1) +</span>
  <span class='fu'>scale_x_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_y_log10</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_color_OkabeIto</span><span class='op'>(</span>order <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>8</span><span class='op'>)</span>, use_black <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_minimal_grid.html'>theme_minimal_grid</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>coord_fixed</span><span class='op'>(</span>clip <span class='op'>=</span> <span class='st'>'off'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="inspect-s1-sequencing-results_files/figure-html5/unnamed-chunk-20-1.png" width="672" /></p>
</div>
<p>Note, in this case the error bars are a bit distracting b/c we don’t have replicates for many of the points, so I dropped them.</p>
<p>here we can see that extraction protocol 2 shows saturation; while protocol 1 may show a drop off at low biomass; these were perhaps canceling each other out in our earlier plot combining the two.</p>
<h2 id="reads-1">Reads</h2>
<p>It is somewhat strange that the couple samples with substantial DNA failed.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>%&gt;%</span> 
  <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>read_depth</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>select</span><span class='op'>(</span><span class='va'>dna_sample_id</span>, <span class='va'>dilution_power</span>, <span class='va'>dna_conc</span>, <span class='va'>qpcr_copies</span>, <span class='va'>comments</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 5 × 5
#&gt;   dna_sample_id dilution_power dna_conc qpcr_copies comments          
#&gt;   &lt;chr&gt;                  &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;             
#&gt; 1 F1-D0-4                    0    1.03      7275245 &quot;&quot;                
#&gt; 2 F2-D2-3                    2    0         1095264 &quot;Insufficient mat…
#&gt; 3 M2N-D0-2                   0   54.5     230381263 &quot;&quot;                
#&gt; 4 M3F-D1-2                   1    3.25     28543112 &quot;&quot;                
#&gt; 5 M3F-D2-2                   2    0.564     6164659 &quot;Marginal materia…</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sam1</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>arrange</span><span class='op'>(</span><span class='va'>read_depth</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>select</span><span class='op'>(</span><span class='va'>dna_sample_id</span>, <span class='va'>read_depth</span>, <span class='va'>dna_conc</span>, <span class='va'>qpcr_copies</span>, <span class='va'>comments</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span>n<span class='op'>=</span><span class='fl'>30</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 94 × 5
#&gt;    dna_sample_id read_depth dna_conc qpcr_copies comments             
#&gt;    &lt;chr&gt;              &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;                
#&gt;  1 M1N-D2-3           41586    0          964017 &quot;Insufficient materi…
#&gt;  2 F1-D2-2            48082    0               0 &quot;Insufficient materi…
#&gt;  3 F3-D2-1            77464    0               0 &quot;Insufficient materi…
#&gt;  4 F1-D2-4           161897    0          292137 &quot;Insufficient materi…
#&gt;  5 F1-D2-3           268021    0          230499 &quot;Insufficient materi…
#&gt;  6 F1-D2-1           281202    0          243013 &quot;Insufficient materi…
#&gt;  7 M3F-D2-1          312964    0          724533 &quot;Insufficient materi…
#&gt;  8 T1-D0-1           402541    0          687679 &quot;Insufficient materi…
#&gt;  9 F2-D2-1           556926    0               0 &quot;Insufficient materi…
#&gt; 10 F1-D1-4           603241    0.311     3106346 &quot;Marginal material f…
#&gt; 11 M2F-D0-4         1053907  223.      199563228 &quot;&quot;                   
#&gt; 12 M1N-D0-4         1113787   41.6     235422296 &quot;&quot;                   
#&gt; 13 F2-D2-4          1147633    0.061     1314673 &quot;Insufficient materi…
#&gt; 14 F2-D2-2          1452843    0.09      1495039 &quot;Insufficient materi…
#&gt; 15 M2N-D2-1         1719314    0.164     1596785 &quot;Insufficient materi…
#&gt; 16 F1-D1-2          1862742    0         1192657 &quot;Insufficient materi…
#&gt; 17 M1F-D2-3         2172356    0.232     2017477 &quot;Insufficient materi…
#&gt; 18 F3-D0-2          2638823    0.174     2812040 &quot;Insufficient materi…
#&gt; 19 F3-D1-2          2796358    0.249     2462811 &quot;Insufficient materi…
#&gt; 20 F1-D1-3          2904362    0         1753585 &quot;Insufficient materi…
#&gt; 21 M3N-D2-1         3117998    0.257     1901696 &quot;Insufficient materi…
#&gt; 22 M2N-D2-3         3212898    0.408     3107850 &quot;Marginal material f…
#&gt; 23 F1-D1-1          3323622    0.496     2802958 &quot;Marginal material f…
#&gt; 24 M2N-D0-4         3529736   50.4     267735590 &quot;&quot;                   
#&gt; 25 M1F-D2-1         3721038    0.59      5203004 &quot;Marginal material f…
#&gt; 26 M1F-D2-4         3722293    0.3       2933430 &quot;Insufficient materi…
#&gt; 27 M1F-D0-1         3790770  102.      555483111 &quot;&quot;                   
#&gt; 28 M1F-D2-2         3817919    0.523     4519188 &quot;Marginal material f…
#&gt; 29 M3N-D1-2         3908574    6.11     39374129 &quot;&quot;                   
#&gt; 30 M1N-D2-4         4092656    1.17      7744146 &quot;&quot;                   
#&gt; # … with 64 more rows</code></pre>
</div>
<p>Also interesting just how unrelated read depth seems to be with qpcr copies.</p>
<p>Generally though it makes sense that these are primarily the samples with low concentrations, and it seems like the failure rate in terms of reads below 600K is generally low - its just the three samples F1-D0-4, M2N-D0-2, and M3F-D1-2 that had sufficient DNA (by picogreen) but weren’t successfully sequenced; all others with sufficient DNA got 1M+ reads.</p>
<h2 class="appendix" id="session-info">Session info</h2>
<details>
<summary>
Click for session info
</summary>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/sessioninfo/man/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; ─ Session info ──────────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 4.1.1 (2021-08-10)
#&gt;  os       Arch Linux                  
#&gt;  system   x86_64, linux-gnu           
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       America/New_York            
#&gt;  date     2021-10-01                  
#&gt; 
#&gt; ─ Packages ──────────────────────────────────────────────────────────────────────
#&gt;  package          * version    date       lib
#&gt;  ade4               1.7-17     2021-06-17 [1]
#&gt;  ape                5.5        2021-04-25 [1]
#&gt;  assertthat         0.2.1      2019-03-21 [1]
#&gt;  backports          1.2.1      2020-12-09 [1]
#&gt;  Biobase            2.52.0     2021-05-19 [1]
#&gt;  BiocGenerics       0.38.0     2021-05-19 [1]
#&gt;  biomformat         1.20.0     2021-05-19 [1]
#&gt;  Biostrings         2.60.1     2021-06-06 [1]
#&gt;  bit                4.0.4      2020-08-04 [1]
#&gt;  bit64              4.0.5      2020-08-30 [1]
#&gt;  bitops             1.0-7      2021-04-24 [1]
#&gt;  broom              0.7.9      2021-07-27 [1]
#&gt;  bslib              0.2.5.1    2021-05-18 [1]
#&gt;  cellranger         1.1.0      2016-07-27 [1]
#&gt;  cli                3.0.1      2021-07-17 [1]
#&gt;  cluster            2.1.2      2021-04-17 [2]
#&gt;  codetools          0.2-18     2020-11-04 [2]
#&gt;  colorblindr        0.1.0      2021-08-27 [1]
#&gt;  colorspace         2.0-2      2021-08-11 [1]
#&gt;  cowplot          * 1.1.1      2021-08-27 [1]
#&gt;  crayon             1.4.1      2021-02-08 [1]
#&gt;  data.table         1.14.0     2021-02-21 [1]
#&gt;  DBI                1.1.1      2021-01-15 [1]
#&gt;  dbplyr             2.1.1      2021-04-06 [1]
#&gt;  digest             0.6.27     2020-10-24 [1]
#&gt;  distill            1.2        2021-01-13 [1]
#&gt;  distributional     0.2.2      2021-02-02 [1]
#&gt;  downlit            0.2.1      2020-11-04 [1]
#&gt;  dplyr            * 1.0.7      2021-06-18 [1]
#&gt;  ellipsis           0.3.2      2021-04-29 [1]
#&gt;  evaluate           0.14       2019-05-28 [1]
#&gt;  fansi              0.5.0      2021-05-25 [1]
#&gt;  farver             2.1.0      2021-02-28 [1]
#&gt;  forcats          * 0.5.1      2021-01-27 [1]
#&gt;  foreach            1.5.1      2020-10-15 [1]
#&gt;  fs               * 1.5.0      2020-07-31 [1]
#&gt;  generics           0.1.0      2020-10-31 [1]
#&gt;  GenomeInfoDb       1.28.1     2021-07-01 [1]
#&gt;  GenomeInfoDbData   1.2.6      2021-05-31 [1]
#&gt;  ggdist           * 3.0.0      2021-07-19 [1]
#&gt;  ggplot2          * 3.3.5      2021-06-25 [1]
#&gt;  glue               1.4.2      2020-08-27 [1]
#&gt;  gtable             0.3.0      2019-03-25 [1]
#&gt;  haven              2.4.1      2021-04-23 [1]
#&gt;  here             * 1.0.1      2020-12-13 [1]
#&gt;  highr              0.9        2021-04-16 [1]
#&gt;  hms                1.1.0      2021-05-17 [1]
#&gt;  htmltools          0.5.1.1    2021-01-22 [1]
#&gt;  httr               1.4.2      2020-07-20 [1]
#&gt;  igraph             1.2.6      2020-10-06 [1]
#&gt;  import             1.2.0      2020-09-24 [1]
#&gt;  IRanges            2.26.0     2021-05-19 [1]
#&gt;  iterators          1.0.13     2020-10-15 [1]
#&gt;  janitor            2.1.0      2021-01-05 [1]
#&gt;  jquerylib          0.1.4      2021-04-26 [1]
#&gt;  jsonlite           1.7.2      2020-12-09 [1]
#&gt;  knitr              1.33       2021-04-24 [1]
#&gt;  lattice            0.20-44    2021-05-02 [2]
#&gt;  lifecycle          1.0.0      2021-02-15 [1]
#&gt;  lubridate          1.7.10     2021-02-26 [1]
#&gt;  magrittr           2.0.1      2020-11-17 [1]
#&gt;  MASS               7.3-54     2021-05-03 [2]
#&gt;  Matrix             1.3-4      2021-06-01 [2]
#&gt;  metacal            0.2.0.9005 2021-07-31 [1]
#&gt;  mgcv               1.8-36     2021-06-01 [2]
#&gt;  modelr             0.1.8      2020-05-19 [1]
#&gt;  multtest           2.48.0     2021-05-19 [1]
#&gt;  munsell            0.5.0      2018-06-12 [1]
#&gt;  nlme               3.1-152    2021-02-04 [2]
#&gt;  nvimcom          * 0.9-102    2021-08-10 [1]
#&gt;  patchwork        * 1.1.1      2020-12-17 [1]
#&gt;  permute            0.9-5      2019-03-12 [1]
#&gt;  phyloseq           1.36.0     2021-05-19 [1]
#&gt;  pillar             1.6.2      2021-07-29 [1]
#&gt;  pkgconfig          2.0.3      2019-09-22 [1]
#&gt;  plyr               1.8.6      2020-03-03 [1]
#&gt;  png                0.1-7      2013-12-03 [1]
#&gt;  purrr            * 0.3.4      2020-04-17 [1]
#&gt;  R6                 2.5.0      2020-10-28 [1]
#&gt;  Rcpp               1.0.7      2021-07-07 [1]
#&gt;  RCurl              1.98-1.3   2021-03-16 [1]
#&gt;  readr            * 2.0.0      2021-07-20 [1]
#&gt;  readxl             1.3.1      2019-03-13 [1]
#&gt;  reprex             2.0.0      2021-04-02 [1]
#&gt;  reshape2           1.4.4      2020-04-09 [1]
#&gt;  rhdf5              2.36.0     2021-05-19 [1]
#&gt;  rhdf5filters       1.4.0      2021-05-19 [1]
#&gt;  Rhdf5lib           1.14.2     2021-07-06 [1]
#&gt;  rJava              1.0-5      2021-09-24 [1]
#&gt;  rlang              0.4.11     2021-04-30 [1]
#&gt;  rmarkdown        * 2.9        2021-06-15 [1]
#&gt;  rprojroot          2.0.2      2020-11-15 [1]
#&gt;  rstudioapi         0.13       2020-11-12 [1]
#&gt;  rvest              1.0.1      2021-07-26 [1]
#&gt;  S4Vectors          0.30.0     2021-05-19 [1]
#&gt;  sass               0.4.0      2021-05-12 [1]
#&gt;  scales             1.1.1      2020-05-11 [1]
#&gt;  sessioninfo        1.1.1      2018-11-05 [1]
#&gt;  snakecase          0.11.0     2019-05-25 [1]
#&gt;  stringi            1.7.3      2021-07-16 [1]
#&gt;  stringr          * 1.4.0      2019-02-10 [1]
#&gt;  survival           3.2-11     2021-04-26 [2]
#&gt;  tabulizer          0.2.2      2018-06-07 [1]
#&gt;  tabulizerjars      1.0.1      2018-05-14 [1]
#&gt;  tibble           * 3.1.3      2021-07-23 [1]
#&gt;  tidyr            * 1.1.3      2021-03-03 [1]
#&gt;  tidyselect         1.1.1      2021-04-30 [1]
#&gt;  tidyverse        * 1.3.1      2021-04-15 [1]
#&gt;  tzdb               0.1.2      2021-07-20 [1]
#&gt;  useful             1.2.6      2018-10-08 [1]
#&gt;  utf8               1.2.2      2021-07-24 [1]
#&gt;  vctrs              0.3.8      2021-04-29 [1]
#&gt;  vegan              2.5-7      2020-11-28 [1]
#&gt;  vroom              1.5.3      2021-07-14 [1]
#&gt;  withr              2.4.2      2021-04-18 [1]
#&gt;  xfun               0.24       2021-06-15 [1]
#&gt;  xml2               1.3.2      2020-04-23 [1]
#&gt;  XVector            0.32.0     2021-05-19 [1]
#&gt;  yaml               2.2.1      2020-02-01 [1]
#&gt;  zlibbioc           1.38.0     2021-05-19 [1]
#&gt;  source                                 
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.0.2)                         
#&gt;  CRAN (R 4.0.2)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  Github (clauswilke/colorblindr@e6730be)
#&gt;  R-Forge (R 4.1.1)                      
#&gt;  Github (wilkelab/cowplot@555c9ae)      
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.5)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.2)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.2)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.5)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.2)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.2)                         
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  Github (mikemc/metacal@0d54dbf)        
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  local                                  
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.5)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.5)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.1)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.4)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.3)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.5)                         
#&gt;  CRAN (R 4.1.0)                         
#&gt;  CRAN (R 4.0.0)                         
#&gt;  Bioconductor                           
#&gt;  CRAN (R 4.0.0)                         
#&gt;  Bioconductor                           
#&gt; 
#&gt; [1] /home/michael/.local/lib/R/library
#&gt; [2] /usr/lib/R/library</code></pre>
</div>
</details>
<div class="sourceCode" id="cb16"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
