---
title: ""
author: Michael McLaren
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(progress = TRUE, verbose = TRUE)
# Global chunk options
knitr::opts_chunk$set(
  include = FALSE, echo = FALSE,
  warning = TRUE, message = FALSE, 
  fig.width = 7, fig.height = 7,
  collapse = TRUE, comment = "#>",
  #> cache = TRUE, autodep = TRUE,
  cache.comments = FALSE
)
```

```{r}
library(tidyverse)
library(here)
library(speedyseq)

library(data.table)

library(kableExtra)

library(cowplot)
library(patchwork)
theme_set(theme_cowplot())
library(ggridges)
library(ggbeeswarm)
library(ggdist)

library(generics)

library(recipes)
#> library(mniw)
library(fido)
library(metacal)
devtools::load_all(here("rpkg"), export_all = FALSE, helpers = FALSE)
```

```{r}
my_psmelt <- function(physeq) {
  speedyseq::psmelt(physeq, as = "tibble") %>%
    rename(taxon = OTU, sample = Sample, abundance = Abundance)
}
as_tibble.phyloseq <- function(x) {
  x %>% my_psmelt
}
as_tibble.sample_data <- function(x) {
  x %>% as("data.frame") %>% as_tibble(rownames = "sample")
}
as_tibble.taxonomyTable <- function(x) {
  x %>% as("matrix") %>% as_tibble(rownames = "taxon")
}
as_tibble.XStringSet <- function(x) {
  x %>% as.character %>% enframe("taxon", "sequence")
}
my_kbl <- function(x, position = "center", ...) {
  kbl(x, ...) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"),
    full_width = FALSE, position = position)
}
```

```{r load-data, cache = TRUE}
ps <- here("output/phyloseq/2020-11-28-phyloseq.Rds") %>% 
  readRDS
## Adjust sample data
sample_data(ps) <- sample_data(ps) %>% as_tibble %>%
  mutate(
    sample_sum = sample_sums(ps),
    across(extraction_batch, fct_relabel, ~str_c("B", .)),
    across(extraction_protocol, fct_relabel, ~str_c("E", .)),
    protocol = extraction_protocol:center_id,
    batch_protocol = extraction_batch:protocol,
    #> batch_protocol = fct_other(extraction_batch:extraction_protocol:center_id,
    #>   drop = c("B1:E1:A1", "B2:E1:A1"), other_level = "Ref") %>%
    #>   fct_relevel("Ref"),
    # Center-id that combines A1 and S1 into a "Ref" level
    #> center_id_ref = fct_other(center_id, drop = c("A1", "S1"), 
    #>   other_level = "Ref") %>%
    #>   fct_relevel("Ref")
  ) %>%
  sample_data
assignments <- tax_table(ps) %>% 
  as_tibble %>%
  mutate(across(c(organism, strain_group), str_split, "; ")) %>%
  unnest(c(organism, strain_group))
inoculum_species <- assignments %>%
  filter(
    strain_group %in% "Inoculum",
  ) %>%
  pull(taxon)
focal_species <- assignments %>%
  filter(
    strain_group %in% c("Inoculum", "Mouse contaminant"),
    species != "Faecalibacterium prausnitzii",
    species != "Eubacterium rectale",
    species != "Escherichia coli",
  ) %>%
  pull(taxon)
```

```{r}
ps0 <- ps %>%
  prune_samples(sample_sums(.) > 1e3, .) %>%
  subset_samples(specimen_type %in% c("Fecal", "Inoculum"))
# Merge to genus level for Staph and Bacillus contams; species for everything
# else
tax_table(ps0) <- tax_table(ps0) %>% as_tibble %>%
  mutate(
    merged_taxon = case_when(
      genus %in% c("Bacillus", "Staphylococcus") ~ genus,
      TRUE ~ species
    ),
    species = ifelse(species == merged_taxon, species, NA)
  ) %>%
  select(taxon, phylum:species, merged_taxon) %>%
  tax_table
ps1 <- merge_taxa_vec(ps0, c(tax_table(ps0)[,"merged_taxon"]))
taxa_names(ps1) <- c(tax_table(ps1)[,"merged_taxon"])
min(sample_sums(ps1) / sample_sums(ps0))
```


```{r cache = TRUE}
ps0 <- ps %>%
  prune_taxa(focal_taxa, .) %>%
  prune_samples(sample_sums(.) > 1e3, .) %>%
  subset_samples(specimen_type %in% c("Fecal", "Inoculum")) %>%
  tax_glom("species")
taxa_names(ps0) <- tax_table(ps0)[,"species"] %>% c
taxa_names(ps0)
tree <- here("output/strain-data/gtdb-representatives.tree") %>%
  ape::read.tree()
focal_id <- here("output/strain-data/focal-strains.csv") %>%
  read_csv
tree$tip.label <- focal_id$display_name[
  match(
    tree$tip.label,
    focal_id$ref_assembly_accession
  )] %>%
  str_replace(" HS$", "")
ps0 <- merge_phyloseq(ps0, tree)
# Strain metadata (currently from v1 table)
strains <- here("output/strain-data", "v1/strain-pheno-geno-tax.Rds") %>%
  readRDS %>%
  mutate(taxon = ncbi_species_display_name) %>%
  filter(taxon %in% taxa_names(ps0)) %>%
  left_join(tax_table(ps0) %>% as_tibble, by = "taxon") %>%
  mutate(
    across(taxon, factor, levels = taxa_names(ps0)),
    across(gram_stain, str_replace, "Gram", "")
  )
# genome bias: 16s copy numbers and sizes
genome_bias <- strains %>%
  filter(taxon %in% taxa_names(ps0)) %>%
  select(
    taxon,
    ssu_count,
    reference_genome_size
  ) %>%
  mutate(
    across(-c(taxon), center_elts)
  )
sam <- sample_data(ps0) %>% as_tibble
tree <- phy_tree(ps0)
```

```{r, eval = FALSE, include = FALSE}
strains %>%
  select(gram_stain, phylum, taxon) %>% 
  arrange(desc(taxon)) %>%
  my_kbl
```

```{r tree-plot, include = TRUE, fig.dim = c(9,5)}
lbls <- strains %>% 
  arrange(taxon) %>%
  transmute(label = str_glue("({gram_stain}) {phylum} > {taxon}")) %>%
  pull(label)
tree0 <- tree
tree0$tip.label <- lbls
plot(tree0)
rm(lbls, tree0)
```


## Background

### Extraction protocols

#### Extraction Protocol 1 (Fast DNA)

- [QIAamp Fast DNA Stool Mini Kit](https://www.qiagen.com/us/en/products/discovery-and-translational-research/dna-rna-purification/dna-purification/microbial-dna/qiaamp-fast-dna-stool-mini-kit/?clear=true#orderinginformation)
- [Modified protocol from Knudsen et al 2016](https://drive.google.com/file/d/1kMxY33tBbSCDWwB_Dmn6vBXI9xr7yZ3O/view?usp=sharing)
- [Ali's minor modification of Knudson protocol](https://drive.google.com/open?id=0B49snwlXGiOeRlJMZ0xDU3BJamJZMm1NQWo4ay1OQWZPLUhF), using bead beating with Matrix E Bead beating tubes instead of the TissueLyser step.

#### Extraction Protocol 2 (DNeasy)

- Kit: [DNeasy Blood and Tissue Kit](https://www.qiagen.com/us/products/discovery-and-translational-research/dna-rna-purification/dna-purification/genomic-dna/dneasy-blood-and-tissue-kit/)
- Protocol: Quick start protocol w/ minor modification to elution step to increase concentration

### Relative abundances

```{r}
ps0 <- ps %>% 

  tax_glom("species")
summary(sample_sums(ps.species) / sample_sums(ps))

pstb <- my_psmelt(ps.species) %>%
  with_groups(sample, mutate,
    proportion = close_elts(abundance),
    clr = metacal::clr(proportion + 1e-4),
    clr2 = metacal::clr(proportion + 1e-4, base = 2),
  )
```
```{r}
p1 <- pstb %>%
  ggplot(aes(y = taxon, x = proportion, color = protocol)) +
  facet_grid(specimen_type ~ extraction_batch) +
  #> geom_quasirandom(groupOnX = FALSE) +
  stat_pointinterval(.width = 0.9, position = position_dodge(width = 0.7)) +
  scale_x_continuous(
    trans = scales::pseudo_log_trans(1e-3),
    breaks = c(1e-3, 1e-2, 3e-2, 1e-1, 0.3)
  ) +
  theme_minimal_hgrid()
p1
```

## Appendix

### DNA yields

DNA yield differs by protocol and type

```{r dna-conc}
sam %>%
  filter(specimen_type %in% c("fecal", "inoculum", "T. mobilis")) %>%
  ggplot(aes(specimen_type, dna_conc, color = extraction_protocol)) +
  geom_boxplot() +
  geom_quasirandom(dodge.width = 0.75) +
  # scale_y_sqrt(breaks = c(0.1, 1, 5, 10, 20, 30)) +
  scale_y_log10() +
  facet_wrap(~extraction_batch,
    labeller = as_labeller(function(x) paste("Batch", x))
  ) +
  scale_color_brewer(type = "qual") +
  labs(
    x = "Specimen type", y = "DNA concentration (ng/uL)",
    color = "Extraction Protocol",
    title = "DNA concentration by protocol, specimen type, and batch"
  ) +
  theme(
    legend.position = "bottom",
    panel.spacing = unit(0.03, "npc")
  )
```

- Protocol 1 gave higher DNA yields on fecal samples, but Protocol 2 gave higher yields on inoculum samples.

Could this have something to do with the bias in favor of G+'s in E2?

See file:///home/michael/research/vivo-vitro/main/analysis/2020-03-23-dna-concentrations.html for more thoughts.

