<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>


  <!--radix_placeholder_meta_tags-->
  <title>DADA2 pipeline on the As data for the 2021 experiment</title>

  <meta property="description" itemprop="description" content="This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment."/>


  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2022-02-17"/>
  <meta property="article:created" itemprop="dateCreated" content="2022-02-17"/>
  <meta name="article:author" content="Michael R. McLaren"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="DADA2 pipeline on the As data for the 2021 experiment"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="DADA2 pipeline on the As data for the 2021 experiment"/>
  <meta property="twitter:description" content="This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment."/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","categories","date","output"]}},"value":[{"type":"character","attributes":{},"value":["DADA2 pipeline on the As data for the 2021 experiment"]},{"type":"character","attributes":{},"value":["This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Michael R. McLaren"]}]}]},{"type":"character","attributes":{},"value":["2021 experiment","bioinformatics"]},{"type":"character","attributes":{},"value":["2022-02-17"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-12-1.png","dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-13-1.png","dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-19-1.png","dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-19-2.png"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          return "<p>" + $('#ref-' + ref).html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/header-attrs-2.11/header-attrs.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/popper-2.6.0/popper.min.js"></script>
  <link href="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"DADA2 pipeline on the As data for the 2021 experiment","description":"This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment.","authors":[{"author":"Michael R. McLaren","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2022-02-17T00:00:00.000-05:00","citationText":"McLaren, 2022"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>DADA2 pipeline on the As data for the 2021 experiment</h1>
<!--radix_placeholder_categories-->
<div class="dt-tags">
<div class="dt=tag">2021 experiment</div>
<div class="dt=tag">bioinformatics</div>
</div>
<!--/radix_placeholder_categories-->
<p><p>This document analyzes the amplicon data from the A2 sequencing center for the 2021 experiment following the workflow used for the 2019 experiment.</p></p>
</div>

<div class="d-byline">
  Michael R. McLaren  
  
<br/>2022-02-17
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#setup">Setup</a></li>
<li><a href="#check-sequences">Check sequences</a></li>
<li><a href="#pick-truncation-params">Pick truncation params</a></li>
<li><a href="#filter-and-trim">Filter and trim</a></li>
<li><a href="#learn-the-error-rates">Learn the error rates</a></li>
<li><a href="#sample-inference">Sample inference</a>
<ul>
<li><a href="#default-omega_c">Default <code>OMEGA_C</code></a></li>
<li><a href="#omega_c-2"><code>OMEGA_C = 2</code></a></li>
</ul></li>
<li><a href="#session-info">Session info</a></li>
</ul>
</nav>
</div>
<p>We follow the workflow in ‘analysis/2020-11-01-dada2/a2-dada2.html’ that was used for the 2019 experiment.</p>
<h2 id="setup">Setup</h2>
<p>Load non-bioinformatic libraries,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; here() starts at /home/michael/research/vivo-vitro</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; ── Attaching packages ─────────────────────────────────── tidyverse 1.3.1 ──
#&gt; ✔ ggplot2 3.3.5     ✔ purrr   0.3.4
#&gt; ✔ tibble  3.1.6     ✔ dplyr   1.0.8
#&gt; ✔ tidyr   1.2.0     ✔ stringr 1.4.0
#&gt; ✔ readr   2.1.1     ✔ forcats 0.5.1
#&gt; ── Conflicts ────────────────────────────────────── tidyverse_conflicts() ──
#&gt; ✖ dplyr::filter() masks stats::filter()
#&gt; ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://fs.r-lib.org'>fs</a></span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Next, I will install and load the previous DADA2 version that I used for the 2020-11-01 analysis of the 2019 data: version 1.18.0 from Bioconductor, which corresponds to the v1.18 tag (a20a676) on GitHub (<a href="https://github.com/benjjneb/dada2/releases/tag/v1.18" class="uri">https://github.com/benjjneb/dada2/releases/tag/v1.18</a>). (Only run if not previously done.)</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>'lib'</span>, <span class='st'>'R'</span>, <span class='st'>'library-for-dada2-1.18.0'</span><span class='op'>)</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://fs.r-lib.org/reference/file_access.html'>dir_exists</a></span><span class='op'>(</span><span class='va'>dr</span><span class='op'>)</span><span class='op'>)</span> 
  <span class='fu'><a href='https://fs.r-lib.org/reference/create.html'>dir_create</a></span><span class='op'>(</span><span class='va'>dr</span><span class='op'>)</span>
<span class='fu'>devtools</span><span class='fu'>::</span><span class='fu'><a href='https://devtools.r-lib.org/reference/dev_mode.html'>dev_mode</a></span><span class='op'>(</span><span class='cn'>TRUE</span>, <span class='va'>dr</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; ✔ Dev mode: ON</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/packageDescription.html'>packageVersion</a></span><span class='op'>(</span><span class='st'>"dada2"</span><span class='op'>)</span> <span class='op'>!=</span> <span class='st'>"1.18.0"</span><span class='op'>)</span>
  <span class='fu'>remotes</span><span class='fu'>::</span><span class='fu'><a href='https://remotes.r-lib.org/reference/install_github.html'>install_github</a></span><span class='op'>(</span><span class='st'>'benjjneb/dada2@v1.18'</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Load DADA2 and functions from bioinformatics libraries</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/stopifnot.html'>stopifnot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/packageDescription.html'>packageVersion</a></span><span class='op'>(</span><span class='st'>"dada2"</span><span class='op'>)</span> <span class='op'>==</span> <span class='st'>"1.18.0"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://benjjneb.github.io/dada2/'>dada2</a></span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Loading required package: Rcpp</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r">
<span class='fu'>import</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/import/man/importfunctions.html'>from</a></span><span class='op'>(</span><span class='va'>Biostrings</span>, <span class='va'>complement</span>, <span class='va'>reverseComplement</span>, <span class='va'>width</span>, <span class='va'>DNAString</span>,
  <span class='va'>readDNAStringSet</span>, <span class='va'>subseq</span><span class='op'>)</span>
<span class='fu'>import</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/import/man/importfunctions.html'>from</a></span><span class='op'>(</span><span class='va'>DECIPHER</span>, <span class='va'>TrimDNA</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Path for saving dada2 results:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>results_path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"2021"</span>, <span class='st'>"a2"</span>, <span class='st'>"dada2"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='va'>results_path</span>, recursive <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning in dir.create(results_path, recursive = TRUE): &#39;/home/michael/
#&gt; research/vivo-vitro/output/2021/a2/dada2&#39; already exists</code></pre>
</div>
<p>Path to the fastq files:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>reads_path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"2021"</span>, <span class='st'>"a2"</span>, <span class='st'>"reads"</span><span class='op'>)</span>
<span class='fu'><a href='https://fs.r-lib.org/reference/dir_ls.html'>dir_ls</a></span><span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"raw"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>path_file</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>head</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;F1-D0-1_S10_L001_R1_001.fastq.gz&quot;
#&gt; [2] &quot;F1-D0-1_S10_L001_R2_001.fastq.gz&quot;
#&gt; [3] &quot;F1-D0-2_S89_L001_R1_001.fastq.gz&quot;
#&gt; [4] &quot;F1-D0-2_S89_L001_R2_001.fastq.gz&quot;
#&gt; [5] &quot;F1-D0-3_S17_L001_R1_001.fastq.gz&quot;
#&gt; [6] &quot;F1-D0-3_S17_L001_R2_001.fastq.gz&quot;</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://fs.r-lib.org/reference/dir_ls.html'>dir_ls</a></span><span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"raw"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>path_file</span>
</code></pre>
</div>
</div>
<p>Let’s get a data frame with the sample names, paths to the raw reads, and paths to the (to-be-created) filtered reads. Note the creation of DNA sample names from the sequence files: I revert the hyphens back to underscores, make the names of the A2 control samples lowercase, and prepend 2021 to ensure uniqueness and distinguishability from the 2019 DNA samples. The sequencing center supplied the reads that could not be demultiplxed in files prefixed ‘Undetermined’; I will not include these in the analysis.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>ftb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://tibble.tidyverse.org/reference/tibble.html'>tibble</a></span><span class='op'>(</span>path <span class='op'>=</span> <span class='fu'><a href='https://fs.r-lib.org/reference/dir_ls.html'>dir_ls</a></span><span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"raw"</span><span class='op'>)</span>, glob <span class='op'>=</span> <span class='st'>"*.fastq.gz"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>
    read_direction <span class='op'>=</span> <span class='fu'><a href='https://dplyr.tidyverse.org/reference/case_when.html'>case_when</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_detect.html'>str_detect</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"_R1_"</span><span class='op'>)</span> <span class='op'>~</span> <span class='st'>"R1"</span>,
      <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_detect.html'>str_detect</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='st'>"_R2_"</span><span class='op'>)</span> <span class='op'>~</span> <span class='st'>"R2"</span>
    <span class='op'>)</span>,
    fastq_sample_id <span class='op'>=</span> <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_extract.html'>str_extract</a></span><span class='op'>(</span><span class='va'>path</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>path_file</span>, <span class='st'>"[^_]+"</span><span class='op'>)</span>,
    dna_sample_id <span class='op'>=</span> <span class='fu'><a href='https://dplyr.tidyverse.org/reference/case_when.html'>case_when</a></span><span class='op'>(</span>
      <span class='va'>fastq_sample_id</span> <span class='op'><a href='https://rdrr.io/r/base/match.html'>%in%</a></span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'MockZymoPos'</span>, <span class='st'>'WaterNeg'</span><span class='op'>)</span> <span class='op'>~</span>
        <span class='fu'><a href='https://stringr.tidyverse.org/reference/case.html'>str_to_lower</a></span><span class='op'>(</span><span class='va'>fastq_sample_id</span><span class='op'>)</span>,
      <span class='cn'>TRUE</span> <span class='op'>~</span> <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_replace.html'>str_replace_all</a></span><span class='op'>(</span><span class='va'>fastq_sample_id</span>, <span class='st'>"-"</span>, <span class='st'>"_"</span><span class='op'>)</span>,
      <span class='op'>)</span>,
    dna_sample_id <span class='op'>=</span> <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_c.html'>str_c</a></span><span class='op'>(</span><span class='st'>"2021_"</span>, <span class='va'>dna_sample_id</span><span class='op'>)</span>,
    sample_id <span class='op'>=</span> <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_c.html'>str_c</a></span><span class='op'>(</span><span class='st'>"A2_"</span>, <span class='va'>dna_sample_id</span><span class='op'>)</span>,
    path_filt <span class='op'>=</span> <span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"filtered"</span>, 
      <span class='fu'><a href='https://stringr.tidyverse.org/reference/str_glue.html'>str_glue</a></span><span class='op'>(</span><span class='st'>"{fastq_sample_id}_{read_direction}_filt.fastq.gz"</span><span class='op'>)</span>
    <span class='op'>)</span>
  <span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://tidyr.tidyverse.org/reference/pivot_wider.html'>pivot_wider</a></span><span class='op'>(</span>names_from <span class='op'>=</span> <span class='va'>read_direction</span>, values_from <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>path</span>, <span class='va'>path_filt</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>fastq_sample_id</span> <span class='op'>!=</span> <span class='st'>'Undetermined'</span><span class='op'>)</span>
<span class='co'># Check</span>
<span class='va'>ftb</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>head</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/across.html'>across</a></span><span class='op'>(</span><span class='fu'><a href='https://tidyselect.r-lib.org/reference/starts_with.html'>starts_with</a></span><span class='op'>(</span><span class='st'>"path"</span><span class='op'>)</span>, <span class='va'>path_file</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>glimpse</span>
</code></pre>
</div>
<pre><code>#&gt; Rows: 6
#&gt; Columns: 7
#&gt; $ fastq_sample_id &lt;chr&gt; &quot;F1-D0-1&quot;, &quot;F1-D0-2&quot;, &quot;F1-D0-3&quot;, &quot;F1-D0-4&quot;, …
#&gt; $ dna_sample_id   &lt;chr&gt; &quot;2021_F1_D0_1&quot;, &quot;2021_F1_D0_2&quot;, &quot;2021_F1_D0_…
#&gt; $ sample_id       &lt;chr&gt; &quot;A2_2021_F1_D0_1&quot;, &quot;A2_2021_F1_D0_2&quot;, &quot;A2_20…
#&gt; $ path_R1         &lt;chr&gt; &quot;F1-D0-1_S10_L001_R1_001.fastq.gz&quot;, &quot;F1-D0-2…
#&gt; $ path_R2         &lt;chr&gt; &quot;F1-D0-1_S10_L001_R2_001.fastq.gz&quot;, &quot;F1-D0-2…
#&gt; $ path_filt_R1    &lt;chr&gt; &quot;F1-D0-1_R1_filt.fastq.gz&quot;, &quot;F1-D0-2_R1_filt…
#&gt; $ path_filt_R2    &lt;chr&gt; &quot;F1-D0-1_R2_filt.fastq.gz&quot;, &quot;F1-D0-2_R2_filt…</code></pre>
</div>
<h2 id="check-sequences">Check sequences</h2>
<p>The <a href="https://support.illumina.com/documents/documentation/chemistry_documentation/16s/16s-metagenomic-library-prep-guide-15044223-b.pdf">Illumina manual</a> lists the primer sequences as</p>
<blockquote>
<p>16S Amplicon PCR Forward Primer = 5’ TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCTACGGGNGGCWGCAG 16S Amplicon PCR Reverse Primer = 5’ GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGACTACHVGGGTATCTAATCC</p>
</blockquote>
<p>and states that the sequenced region is V3-V4 and approximately 460 bp long.</p>
<p>Let’s check to see if the same substrings of the primer sequences as last time are present.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>reads</span> <span class='op'>&lt;-</span> <span class='va'>ftb</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> 
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>dna_sample_id</span> <span class='op'>==</span> <span class='st'>"2021_M3F_D0_1"</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> 
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='va'>path_R1</span>, R2 <span class='op'>=</span> <span class='va'>path_R2</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='va'>c</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>readDNAStringSet</span>, format <span class='op'>=</span> <span class='st'>"fastq"</span><span class='op'>)</span>
<span class='va'>reads</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>subseq</span>, end <span class='op'>=</span> <span class='fl'>30</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; $R1
#&gt; DNAStringSet object of length 210056:
#&gt;          width seq                                names               
#&gt;      [1]    30 CCTACGGGGGGCTGCAGTGAGGAATATTGG     M04754:191:000000...
#&gt;      [2]    30 CCTACGGGCGGCTGCAGTGGGGAATATTGC     M04754:191:000000...
#&gt;      [3]    30 CCTACGGGTGGCAGCAGTAGGGAATCTTCC     M04754:191:000000...
#&gt;      [4]    30 CCTACGGGTGGCAGCAGTGAGGAATATTGG     M04754:191:000000...
#&gt;      [5]    30 CCTACGGGAGGCAGCAGTGAGGAATATTGG     M04754:191:000000...
#&gt;      ...   ... ...
#&gt; [210052]    30 CCTACGGGTGGCTGCAGTGAGGAATATTGG     M04754:191:000000...
#&gt; [210053]    30 CCTACGGGAGGCTGCAGTCGAGAATCATTC     M04754:191:000000...
#&gt; [210054]    30 CCTACGGGTGGCTGCAGTGGGGAATATTGG     M04754:191:000000...
#&gt; [210055]    30 CCTACGGGTGGCAGCAGTGGGGAATATTGC     M04754:191:000000...
#&gt; [210056]    30 CCTACGGGTGGCAGCAGTGAGGAATATTGG     M04754:191:000000...
#&gt; 
#&gt; $R2
#&gt; DNAStringSet object of length 210056:
#&gt;          width seq                                names               
#&gt;      [1]    30 GACTACTCGGGTATCTAATCCTGTTTGATA     M04754:191:000000...
#&gt;      [2]    30 GACTACTGGGGTATCTAATCCTGTTTGCTC     M04754:191:000000...
#&gt;      [3]    30 GGATCACTTGGGCACTCAACCCTGTTTGTC     M04754:191:000000...
#&gt;      [4]    30 GACTACACGGGTATCAACTCCGTTTGACAC     M04754:191:000000...
#&gt;      [5]    30 GCCACCTGGGTACTCAACTCCGTTTGACAC     M04754:191:000000...
#&gt;      ...   ... ...
#&gt; [210052]    30 GANTANCGGGGTATCTAATCCTGTTTGATA     M04754:191:000000...
#&gt; [210053]    30 GANTANTAGGGTATCTAATCCCTTTCGCTC     M04754:191:000000...
#&gt; [210054]    30 GANTANCAGGGTATCTAATCCTGTTTGATA     M04754:191:000000...
#&gt; [210055]    30 GANTANTCGGGTATCTAATCCTGTTTGCTA     M04754:191:000000...
#&gt; [210056]    30 GANTANCAGGGTATCTAATCCTGTTTGATA     M04754:191:000000...</code></pre>
</div>
<p>We can see that as before, the following parts of the described primer sequences arepresent and need to be trimmed:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>primers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='st'>"CCTACGGGNGGCWGCAG"</span>, R2 <span class='op'>=</span> <span class='st'>"GACTACHVGGGTATCTAATCC"</span><span class='op'>)</span>
<span class='va'>trim_left</span> <span class='op'>&lt;-</span> <span class='va'>primers</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map_int</a></span><span class='op'>(</span><span class='va'>nchar</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='va'>print</span>
</code></pre>
</div>
<pre><code>#&gt; R1 R2 
#&gt; 17 21</code></pre>
</div>
<h2 id="pick-truncation-params">Pick truncation params</h2>
<p>Previously, I determined that the length of the target region (excluding primer sequences) in our reference genomes was 420-430 bp. Also, I ran FIGARO on the 2019 data to pick the following truncation and maxEE parameters,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>target_len</span> <span class='op'>&lt;-</span> <span class='fl'>430</span>
<span class='va'>trunc_len</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='fl'>276</span>, R2 <span class='op'>=</span> <span class='fl'>222</span><span class='op'>)</span>
<span class='va'>max_ee</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='fl'>4</span>, R2 <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='co'># should overlap by at least</span>
<span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>trunc_len</span><span class='op'>)</span> <span class='op'>-</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>trim_left</span><span class='op'>)</span> <span class='op'>-</span> <span class='va'>target_len</span>
</code></pre>
</div>
<pre><code>#&gt; [1] 30</code></pre>
</div>
<p>Here we’ll aim to use the same parameters. Let’s check some quality profiles to ensure these truncation lengths still look good,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>42</span><span class='op'>)</span>
<span class='va'>idx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>ftb</span><span class='op'>)</span>, <span class='fl'>9</span><span class='op'>)</span>
<span class='va'>idx</span>
</code></pre>
</div>
<pre><code>#&gt; [1] 49 65 25 74 18 96 47 24 71</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>ftb</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://dplyr.tidyverse.org/reference/slice.html'>slice</a></span><span class='op'>(</span><span class='va'>idx</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://dplyr.tidyverse.org/reference/pull.html'>pull</a></span><span class='op'>(</span><span class='va'>sample_id</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;A2_2021_M1N_D1_3&quot; &quot;A2_2021_M2F_D2_3&quot; &quot;A2_2021_F3_D0_1&quot; 
#&gt; [4] &quot;A2_2021_M2N_D1_4&quot; &quot;A2_2021_F2_D1_2&quot;  &quot;A2_2021_T1_D0_4&quot; 
#&gt; [7] &quot;A2_2021_M1N_D1_1&quot; &quot;A2_2021_F2_D2_4&quot;  &quot;A2_2021_M2N_D1_1&quot;</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qps</span> <span class='op'>&lt;-</span> <span class='va'>ftb</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/slice.html'>slice</a></span><span class='op'>(</span><span class='va'>idx</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='va'>path_R1</span>, R2 <span class='op'>=</span> <span class='va'>path_R2</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='va'>c</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>plotQualityProfile</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = &quot;none&quot;)` instead.
#&gt; `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = &quot;none&quot;)` instead.</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='va'>qps</span><span class='op'>[[</span><span class='st'>"R1"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_hline</a></span><span class='op'>(</span>yintercept <span class='op'>=</span> <span class='fl'>30</span>, color <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_vline</a></span><span class='op'>(</span>xintercept <span class='op'>=</span> <span class='va'>trunc_len</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>, color <span class='op'>=</span> <span class='st'>"darkred"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='st'>"Forward reads"</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='va'>qps</span><span class='op'>[[</span><span class='st'>"R2"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_hline</a></span><span class='op'>(</span>yintercept <span class='op'>=</span> <span class='fl'>30</span>, color <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_vline</a></span><span class='op'>(</span>xintercept <span class='op'>=</span> <span class='va'>trunc_len</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>, color <span class='op'>=</span> <span class='st'>"darkred"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='st'>"Reverse reads"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span>
</code></pre>
</div>
<p><img src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p2</span>
</code></pre>
</div>
<p><img src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggsave.html'>ggsave</a></span><span class='op'>(</span>plot <span class='op'>=</span> <span class='va'>p1</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"quality-profiles-r1.pdf"</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>10</span>, height <span class='op'>=</span> <span class='fl'>10</span>, units <span class='op'>=</span> <span class='st'>"in"</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggsave.html'>ggsave</a></span><span class='op'>(</span>plot <span class='op'>=</span> <span class='va'>p2</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"quality-profiles-r2.pdf"</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>10</span>, height <span class='op'>=</span> <span class='fl'>10</span>, units <span class='op'>=</span> <span class='st'>"in"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>The sequencing quality looks good and profiles are similar to the 2019 experiment, so these truncation parameters should work fine.</p>
<h2 id="filter-and-trim">Filter and trim</h2>
<p>First, delete the filtered reads from any previous runs (if they exist),</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/file_access.html'>dir_exists</a></span><span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"filtered"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
  <span class='fu'><a href='https://fs.r-lib.org/reference/delete.html'>dir_delete</a></span><span class='op'>(</span><span class='fu'><a href='https://fs.r-lib.org/reference/path.html'>path</a></span><span class='op'>(</span><span class='va'>reads_path</span>, <span class='st'>"filtered"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>and set up lists with the paths to the forward and reverse reads (raw and filtered). Note, the file paths are named by sample id so that later DADA2 functions will name the samples by sample id instead of file name.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>path_raw</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>path_R1</span>, R2 <span class='op'>=</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>path_R2</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>set_names</span>, <span class='va'>ftb</span><span class='op'>$</span><span class='va'>sample_id</span><span class='op'>)</span>
<span class='va'>path_filt</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>R1 <span class='op'>=</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>path_filt_R1</span>, R2 <span class='op'>=</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>path_filt_R2</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>set_names</span>, <span class='va'>ftb</span><span class='op'>$</span><span class='va'>sample_id</span><span class='op'>)</span>
<span class='co'># Check</span>
<span class='va'>path_filt</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>tail</span>, <span class='fl'>4</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>names</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; $R1
#&gt; [1] &quot;A2_2021_T1_D0_1&quot; &quot;A2_2021_T1_D0_2&quot; &quot;A2_2021_T1_D0_3&quot;
#&gt; [4] &quot;A2_2021_T1_D0_4&quot;
#&gt; 
#&gt; $R2
#&gt; [1] &quot;A2_2021_T1_D0_1&quot; &quot;A2_2021_T1_D0_2&quot; &quot;A2_2021_T1_D0_3&quot;
#&gt; [4] &quot;A2_2021_T1_D0_4&quot;</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>path_filt</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>tail</span>, <span class='fl'>4</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>path_file</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; $R1
#&gt; [1] &quot;T1-D0-1_R1_filt.fastq.gz&quot; &quot;T1-D0-2_R1_filt.fastq.gz&quot;
#&gt; [3] &quot;T1-D0-3_R1_filt.fastq.gz&quot; &quot;T1-D0-4_R1_filt.fastq.gz&quot;
#&gt; 
#&gt; $R2
#&gt; [1] &quot;T1-D0-1_R2_filt.fastq.gz&quot; &quot;T1-D0-2_R2_filt.fastq.gz&quot;
#&gt; [3] &quot;T1-D0-3_R2_filt.fastq.gz&quot; &quot;T1-D0-4_R2_filt.fastq.gz&quot;</code></pre>
</div>
<p>Then filter and trim.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>out</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/filterAndTrim.html'>filterAndTrim</a></span><span class='op'>(</span>
  <span class='va'>path_raw</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
  <span class='va'>path_raw</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>,
  trimLeft <span class='op'>=</span> <span class='va'>trim_left</span>, truncLen <span class='op'>=</span> <span class='va'>trunc_len</span>, 
  maxN <span class='op'>=</span> <span class='fl'>0</span>, maxEE <span class='op'>=</span> <span class='va'>max_ee</span>, truncQ <span class='op'>=</span> <span class='fl'>2</span>, rm.phix <span class='op'>=</span> <span class='cn'>TRUE</span>, 
  compress <span class='op'>=</span> <span class='cn'>TRUE</span>, multithread <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Creating output directory: /home/michael/research/vivo-vitro/data/2021/a2/reads/filtered</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>out</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;                                  reads.in reads.out
#&gt; F1-D0-1_S10_L001_R1_001.fastq.gz   150600    121742
#&gt; F1-D0-2_S89_L001_R1_001.fastq.gz   160932    128950
#&gt; F1-D0-3_S17_L001_R1_001.fastq.gz   150077    122089
#&gt; F1-D0-4_S39_L001_R1_001.fastq.gz   134714    109607
#&gt; F1-D1-1_S68_L001_R1_001.fastq.gz   151965    122963
#&gt; F1-D1-2_S56_L001_R1_001.fastq.gz   158259    125576</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Save in case of crash downstream</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>out</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"filt-out.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="learn-the-error-rates">Learn the error rates</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>errs</span> <span class='op'>&lt;-</span> <span class='va'>path_filt</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span>
  <span class='fu'><a href='https://purrr.tidyverse.org/reference/map.html'>map</a></span><span class='op'>(</span><span class='va'>learnErrors</span>, multithread <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; 124938492 total bases in 482388 reads from 4 samples will be used for learning the error rates.
#&gt; 121675551 total bases in 605351 reads from 5 samples will be used for learning the error rates.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>errs</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"error-models.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/plotErrors.html'>plotErrors</a></span><span class='op'>(</span><span class='va'>errs</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, nominalQ <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous y-
#&gt; axis</code></pre>
<img src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-19-1.png" width="672" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggsave.html'>ggsave</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"error-profile-r1.pdf"</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>9</span>, height <span class='op'>=</span> <span class='fl'>9</span>, units <span class='op'>=</span> <span class='st'>"in"</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous y-
#&gt; axis</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/plotErrors.html'>plotErrors</a></span><span class='op'>(</span><span class='va'>errs</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, nominalQ <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous y-
#&gt; axis</code></pre>
<img src="dada2-pipeline-on-a2-data-for-the-2021-experiment_files/figure-html5/unnamed-chunk-19-2.png" width="672" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggsave.html'>ggsave</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"error-profile-r2.pdf"</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>9</span>, height <span class='op'>=</span> <span class='fl'>9</span>, units <span class='op'>=</span> <span class='st'>"in"</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous y-
#&gt; axis</code></pre>
</div>
<h2 id="sample-inference">Sample inference</h2>
<p>The following settings for the <code>dada()</code> inference step are used:</p>
<ul>
<li>Pseudo-pooling used to increase sensitivity
<ul>
<li><code>pool = "pseudo"</code> option</li>
<li><code>PSEUDO_PREVALENCE = 4</code> (increased from default of 2 to reduce the number of spurious low-prev ASVs seen in the pilot)</li>
</ul></li>
<li>The inference step (<code>dada()</code>) is run two different ways: The default settings, and with <code>OMEGA_C = 2</code> (which prevents error correction, so that we can assign abundances using just the error-free read-pairs).</li>
</ul>
<h3 id="default-omega_c">Default <code>OMEGA_C</code></h3>
<p>Sample inference w/ <code>dada()</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dadas</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map2.html'>map2</a></span><span class='op'>(</span>
  <span class='va'>path_filt</span>,
  <span class='va'>errs</span>,
  <span class='va'>dada</span>,
  multithread <span class='op'>=</span> <span class='cn'>TRUE</span>, pool <span class='op'>=</span> <span class='st'>"pseudo"</span>,
  PSEUDO_PREVALENCE <span class='op'>=</span> <span class='fl'>4</span>
<span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Sample 1 - 121742 reads in 25575 unique sequences.
#&gt; Sample 2 - 128950 reads in 26403 unique sequences.
#&gt; Sample 3 - 122089 reads in 23363 unique sequences.
#&gt; Sample 4 - 109607 reads in 24030 unique sequences.
#&gt; Sample 5 - 122963 reads in 26667 unique sequences.
#&gt; Sample 6 - 125576 reads in 27192 unique sequences.
#&gt; Sample 7 - 111264 reads in 24130 unique sequences.
#&gt; Sample 8 - 174377 reads in 35475 unique sequences.
#&gt; Sample 9 - 117765 reads in 26204 unique sequences.
#&gt; Sample 10 - 134307 reads in 28686 unique sequences.
#&gt; Sample 11 - 148672 reads in 30167 unique sequences.
#&gt; Sample 12 - 110742 reads in 25222 unique sequences.
#&gt; Sample 13 - 144767 reads in 27835 unique sequences.
#&gt; Sample 14 - 136169 reads in 24914 unique sequences.
#&gt; Sample 15 - 110253 reads in 22058 unique sequences.
#&gt; Sample 16 - 136410 reads in 24539 unique sequences.
#&gt; Sample 17 - 117137 reads in 22606 unique sequences.
#&gt; Sample 18 - 65569 reads in 17913 unique sequences.
#&gt; Sample 19 - 102670 reads in 20851 unique sequences.
#&gt; Sample 20 - 112281 reads in 24814 unique sequences.
#&gt; Sample 21 - 138049 reads in 27394 unique sequences.
#&gt; Sample 22 - 157679 reads in 28636 unique sequences.
#&gt; Sample 23 - 113866 reads in 22445 unique sequences.
#&gt; Sample 24 - 133996 reads in 26824 unique sequences.
#&gt; Sample 25 - 165095 reads in 30527 unique sequences.
#&gt; Sample 26 - 123661 reads in 25576 unique sequences.
#&gt; Sample 27 - 145073 reads in 29058 unique sequences.
#&gt; Sample 28 - 108056 reads in 23024 unique sequences.
#&gt; Sample 29 - 123552 reads in 26434 unique sequences.
#&gt; Sample 30 - 110229 reads in 25884 unique sequences.
#&gt; Sample 31 - 107530 reads in 22463 unique sequences.
#&gt; Sample 32 - 112419 reads in 25782 unique sequences.
#&gt; Sample 33 - 125104 reads in 26440 unique sequences.
#&gt; Sample 34 - 197929 reads in 35784 unique sequences.
#&gt; Sample 35 - 157754 reads in 30631 unique sequences.
#&gt; Sample 36 - 126564 reads in 28762 unique sequences.
#&gt; Sample 37 - 133185 reads in 27801 unique sequences.
#&gt; Sample 38 - 155321 reads in 31368 unique sequences.
#&gt; Sample 39 - 105466 reads in 23726 unique sequences.
#&gt; Sample 40 - 115309 reads in 25816 unique sequences.
#&gt; Sample 41 - 121892 reads in 25785 unique sequences.
#&gt; Sample 42 - 98848 reads in 22943 unique sequences.
#&gt; Sample 43 - 99851 reads in 24810 unique sequences.
#&gt; Sample 44 - 162855 reads in 31337 unique sequences.
#&gt; Sample 45 - 104604 reads in 22669 unique sequences.
#&gt; Sample 46 - 126851 reads in 27724 unique sequences.
#&gt; Sample 47 - 108314 reads in 23521 unique sequences.
#&gt; Sample 48 - 130225 reads in 28055 unique sequences.
#&gt; Sample 49 - 149520 reads in 27458 unique sequences.
#&gt; Sample 50 - 162053 reads in 31445 unique sequences.
#&gt; Sample 51 - 119936 reads in 24325 unique sequences.
#&gt; Sample 52 - 116195 reads in 24568 unique sequences.
#&gt; Sample 53 - 110691 reads in 24432 unique sequences.
#&gt; Sample 54 - 112178 reads in 27041 unique sequences.
#&gt; Sample 55 - 167359 reads in 36355 unique sequences.
#&gt; Sample 56 - 95833 reads in 22172 unique sequences.
#&gt; Sample 57 - 111376 reads in 27055 unique sequences.
#&gt; Sample 58 - 106360 reads in 23371 unique sequences.
#&gt; Sample 59 - 126877 reads in 32229 unique sequences.
#&gt; Sample 60 - 117907 reads in 25255 unique sequences.
#&gt; Sample 61 - 111267 reads in 25950 unique sequences.
#&gt; Sample 62 - 118149 reads in 25339 unique sequences.
#&gt; Sample 63 - 161806 reads in 30974 unique sequences.
#&gt; Sample 64 - 138744 reads in 27896 unique sequences.
#&gt; Sample 65 - 98773 reads in 23452 unique sequences.
#&gt; Sample 66 - 176975 reads in 32963 unique sequences.
#&gt; Sample 67 - 113222 reads in 24923 unique sequences.
#&gt; Sample 68 - 134841 reads in 30698 unique sequences.
#&gt; Sample 69 - 149102 reads in 32356 unique sequences.
#&gt; Sample 70 - 115392 reads in 26864 unique sequences.
#&gt; Sample 71 - 120571 reads in 24846 unique sequences.
#&gt; Sample 72 - 108729 reads in 23280 unique sequences.
#&gt; Sample 73 - 108661 reads in 22716 unique sequences.
#&gt; Sample 74 - 148569 reads in 30846 unique sequences.
#&gt; Sample 75 - 115256 reads in 25053 unique sequences.
#&gt; Sample 76 - 104048 reads in 24693 unique sequences.
#&gt; Sample 77 - 116114 reads in 25206 unique sequences.
#&gt; Sample 78 - 108593 reads in 21090 unique sequences.
#&gt; Sample 79 - 167701 reads in 34161 unique sequences.
#&gt; Sample 80 - 170878 reads in 37831 unique sequences.
#&gt; Sample 81 - 132214 reads in 30000 unique sequences.
#&gt; Sample 82 - 157762 reads in 30346 unique sequences.
#&gt; Sample 83 - 154395 reads in 31945 unique sequences.
#&gt; Sample 84 - 113625 reads in 25282 unique sequences.
#&gt; Sample 85 - 118923 reads in 25640 unique sequences.
#&gt; Sample 86 - 131569 reads in 30861 unique sequences.
#&gt; Sample 87 - 113192 reads in 26263 unique sequences.
#&gt; Sample 88 - 105216 reads in 23463 unique sequences.
#&gt; Sample 89 - 113874 reads in 25122 unique sequences.
#&gt; Sample 90 - 123707 reads in 27819 unique sequences.
#&gt; Sample 91 - 94532 reads in 27867 unique sequences.
#&gt; Sample 92 - 96498 reads in 26033 unique sequences.
#&gt; Sample 93 - 111858 reads in 23220 unique sequences.
#&gt; Sample 94 - 121948 reads in 24379 unique sequences.
#&gt; Sample 95 - 49178 reads in 16475 unique sequences.
#&gt; Sample 96 - 140587 reads in 26137 unique sequences.
#&gt; 
#&gt;    selfConsist step 2Sample 1 - 121742 reads in 21779 unique sequences.
#&gt; Sample 2 - 128950 reads in 23946 unique sequences.
#&gt; Sample 3 - 122089 reads in 24252 unique sequences.
#&gt; Sample 4 - 109607 reads in 18606 unique sequences.
#&gt; Sample 5 - 122963 reads in 26318 unique sequences.
#&gt; Sample 6 - 125576 reads in 22883 unique sequences.
#&gt; Sample 7 - 111264 reads in 19398 unique sequences.
#&gt; Sample 8 - 174377 reads in 30058 unique sequences.
#&gt; Sample 9 - 117765 reads in 24831 unique sequences.
#&gt; Sample 10 - 134307 reads in 28794 unique sequences.
#&gt; Sample 11 - 148672 reads in 28211 unique sequences.
#&gt; Sample 12 - 110742 reads in 22938 unique sequences.
#&gt; Sample 13 - 144767 reads in 27934 unique sequences.
#&gt; Sample 14 - 136169 reads in 27278 unique sequences.
#&gt; Sample 15 - 110253 reads in 20302 unique sequences.
#&gt; Sample 16 - 136410 reads in 25638 unique sequences.
#&gt; Sample 17 - 117137 reads in 22026 unique sequences.
#&gt; Sample 18 - 65569 reads in 13221 unique sequences.
#&gt; Sample 19 - 102670 reads in 19752 unique sequences.
#&gt; Sample 20 - 112281 reads in 24676 unique sequences.
#&gt; Sample 21 - 138049 reads in 28555 unique sequences.
#&gt; Sample 22 - 157679 reads in 29815 unique sequences.
#&gt; Sample 23 - 113866 reads in 21867 unique sequences.
#&gt; Sample 24 - 133996 reads in 24254 unique sequences.
#&gt; Sample 25 - 165095 reads in 31313 unique sequences.
#&gt; Sample 26 - 123661 reads in 25482 unique sequences.
#&gt; Sample 27 - 145073 reads in 26033 unique sequences.
#&gt; Sample 28 - 108056 reads in 20428 unique sequences.
#&gt; Sample 29 - 123552 reads in 26723 unique sequences.
#&gt; Sample 30 - 110229 reads in 20822 unique sequences.
#&gt; Sample 31 - 107530 reads in 21801 unique sequences.
#&gt; Sample 32 - 112419 reads in 21824 unique sequences.
#&gt; Sample 33 - 125104 reads in 25992 unique sequences.
#&gt; Sample 34 - 197929 reads in 44233 unique sequences.
#&gt; Sample 35 - 157754 reads in 32793 unique sequences.
#&gt; Sample 36 - 126564 reads in 24502 unique sequences.
#&gt; Sample 37 - 133185 reads in 26414 unique sequences.
#&gt; Sample 38 - 155321 reads in 34382 unique sequences.
#&gt; Sample 39 - 105466 reads in 20123 unique sequences.
#&gt; Sample 40 - 115309 reads in 24262 unique sequences.
#&gt; Sample 41 - 121892 reads in 26507 unique sequences.
#&gt; Sample 42 - 98848 reads in 21389 unique sequences.
#&gt; Sample 43 - 99851 reads in 18974 unique sequences.
#&gt; Sample 44 - 162855 reads in 30391 unique sequences.
#&gt; Sample 45 - 104604 reads in 18900 unique sequences.
#&gt; Sample 46 - 126851 reads in 26866 unique sequences.
#&gt; Sample 47 - 108314 reads in 22909 unique sequences.
#&gt; Sample 48 - 130225 reads in 22975 unique sequences.
#&gt; Sample 49 - 149520 reads in 33157 unique sequences.
#&gt; Sample 50 - 162053 reads in 33715 unique sequences.
#&gt; Sample 51 - 119936 reads in 23859 unique sequences.
#&gt; Sample 52 - 116195 reads in 24250 unique sequences.
#&gt; Sample 53 - 110691 reads in 23385 unique sequences.
#&gt; Sample 54 - 112178 reads in 32963 unique sequences.
#&gt; Sample 55 - 167359 reads in 30380 unique sequences.
#&gt; Sample 56 - 95833 reads in 17799 unique sequences.
#&gt; Sample 57 - 111376 reads in 20483 unique sequences.
#&gt; Sample 58 - 106360 reads in 20752 unique sequences.
#&gt; Sample 59 - 126877 reads in 22471 unique sequences.
#&gt; Sample 60 - 117907 reads in 25967 unique sequences.
#&gt; Sample 61 - 111267 reads in 21427 unique sequences.
#&gt; Sample 62 - 118149 reads in 22154 unique sequences.
#&gt; Sample 63 - 161806 reads in 31608 unique sequences.
#&gt; Sample 64 - 138744 reads in 27484 unique sequences.
#&gt; Sample 65 - 98773 reads in 18508 unique sequences.
#&gt; Sample 66 - 176975 reads in 37865 unique sequences.
#&gt; Sample 67 - 113222 reads in 21674 unique sequences.
#&gt; Sample 68 - 134841 reads in 25149 unique sequences.
#&gt; Sample 69 - 149102 reads in 27194 unique sequences.
#&gt; Sample 70 - 115392 reads in 21553 unique sequences.
#&gt; Sample 71 - 120571 reads in 21898 unique sequences.
#&gt; Sample 72 - 108729 reads in 20542 unique sequences.
#&gt; Sample 73 - 108661 reads in 21550 unique sequences.
#&gt; Sample 74 - 148569 reads in 28137 unique sequences.
#&gt; Sample 75 - 115256 reads in 21853 unique sequences.
#&gt; Sample 76 - 104048 reads in 20642 unique sequences.
#&gt; Sample 77 - 116114 reads in 20887 unique sequences.
#&gt; Sample 78 - 108593 reads in 21800 unique sequences.
#&gt; Sample 79 - 167701 reads in 35267 unique sequences.
#&gt; Sample 80 - 170878 reads in 31030 unique sequences.
#&gt; Sample 81 - 132214 reads in 24867 unique sequences.
#&gt; Sample 82 - 157762 reads in 35675 unique sequences.
#&gt; Sample 83 - 154395 reads in 33985 unique sequences.
#&gt; Sample 84 - 113625 reads in 22264 unique sequences.
#&gt; Sample 85 - 118923 reads in 25540 unique sequences.
#&gt; Sample 86 - 131569 reads in 29033 unique sequences.
#&gt; Sample 87 - 113192 reads in 23522 unique sequences.
#&gt; Sample 88 - 105216 reads in 23531 unique sequences.
#&gt; Sample 89 - 113874 reads in 23178 unique sequences.
#&gt; Sample 90 - 123707 reads in 24794 unique sequences.
#&gt; Sample 91 - 94532 reads in 27889 unique sequences.
#&gt; Sample 92 - 96498 reads in 26403 unique sequences.
#&gt; Sample 93 - 111858 reads in 23120 unique sequences.
#&gt; Sample 94 - 121948 reads in 23388 unique sequences.
#&gt; Sample 95 - 49178 reads in 30282 unique sequences.
#&gt; Sample 96 - 140587 reads in 25653 unique sequences.
#&gt; 
#&gt;    selfConsist step 2</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>dadas</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"dadas-1.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Check the results:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code>#&gt; dada-class: object describing DADA2 denoising results
#&gt; 369 sequence variants were inferred from 25575 
#&gt; input unique sequences.
#&gt; Key parameters: OMEGA_A = 1e-40, OMEGA_C = 1e-40, BAND_SIZE = 16</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getSequences.html'>getSequences</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>, <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;TGAGGAATATTGGTCAATGGGCGAGAGCCTGAACCAGCCAAGTAGCGTGAAGGATGAAGGCTCTATGGGTCGTAAACTTCTTTTATATGGGAATAAAGTTTTCCACGTGTGGAATTTTGTATGTACCATATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGATTGTTAAGTCAGTTGTGAAAGTTTGCGGCTCA&quot;
#&gt; [2] &quot;TGAGGAATATTGGTCAATGGGCGCAGGCCTGAACCAGCCAAGTAGCGTGAAGGATGACTGCCCTATGGGTTGTAAACTTCTTTTATATGGGAATAAAGTTTTCCACGTGTGGAATTTTGTATGTACCATATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGACAGTTAAGTCAGTTGTGAAAGTTTGCGGCTCA&quot;</code></pre>
</div>
<p>Merge paired reads,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>mergers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/mergePairs.html'>mergePairs</a></span><span class='op'>(</span>
  <span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, 
  <span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>mergers</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"mergers-1.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Construct sequence table</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>seqtab</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/makeSequenceTable.html'>makeSequenceTable</a></span><span class='op'>(</span><span class='va'>mergers</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>seqtab</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"seqtab-1.Rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1]    96 20873</code></pre>
</div>
<p>Inspect distribution of sequence lengths</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nchar.html'>nchar</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getSequences.html'>getSequences</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; 
#&gt;   259   260   267   269   272   276   284   302   323   325   327 
#&gt;    55    20    26     1     3     1     1     1     1     2     2 
#&gt;   328   331   335   343   344   346   348   349   350   351   358 
#&gt;     1     1     7     3     6     1     1     1     5     1     1 
#&gt;   364   365   370   372   373   377   378   379   380   383   384 
#&gt;     1     1     5     1     1     1     1     1     1     7    10 
#&gt;   386   387   388   389   392   393   398   399   400   401   402 
#&gt;     9     1     3     6    18     1     8     2     1    59  1846 
#&gt;   403   404   405   406   407   408   409   411   413   414   415 
#&gt;   460   338    15     8   109   936     5     1     1     4     1 
#&gt;   419   420   421   422   423   424   425   426   427   428   429 
#&gt;     2     9   163 11314    12     1     7    40  4331   959    12 
#&gt;   432   436   438   439   441   444   445   446   447   448 
#&gt;     1     1     1     2     2     3     2     4     3     1</code></pre>
</div>
<p>Remove chimeras (with ‘pooled’ method)</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>seqtab.nochim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/removeBimeraDenovo.html'>removeBimeraDenovo</a></span><span class='op'>(</span><span class='va'>seqtab</span>, method <span class='op'>=</span> <span class='st'>"pooled"</span>,
    multithread <span class='op'>=</span> <span class='cn'>TRUE</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Identified 20018 bimeras out of 20873 input sequences.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"seqtab-nochim-1.Rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1]  96 855</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] 0.7899633</code></pre>
</div>
<p>Track reads through the pipeline,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>getN</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getUniques.html'>getUniques</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>track</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>out</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>getN</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>getN</span><span class='op'>)</span>, 
    <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>mergers</span>, <span class='va'>getN</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"input"</span>, <span class='st'>"filtered"</span>, <span class='st'>"denoised_r1"</span>, <span class='st'>"denoised_r2"</span>, <span class='st'>"merged"</span>,
    <span class='st'>"nochim"</span><span class='op'>)</span>
<span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>sample_id</span>
<span class='va'>track</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://tibble.tidyverse.org/reference/as_tibble.html'>as_tibble</a></span><span class='op'>(</span><span class='va'>track</span>, rownames <span class='op'>=</span> <span class='st'>"sample_id"</span><span class='op'>)</span>
<span class='fu'><a href='https://readr.tidyverse.org/reference/write_delim.html'>write_csv</a></span><span class='op'>(</span><span class='va'>track</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"track-1.csv"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 6 × 7
#&gt;   sample_id       input filtered denoised_r1 denoised_r2 merged nochim
#&gt;   &lt;chr&gt;           &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
#&gt; 1 A2_2021_F1_D0… 150600   121742      121599      121571 119082  84955
#&gt; 2 A2_2021_F1_D0… 160932   128950      128708      128774 126226  94803
#&gt; 3 A2_2021_F1_D0… 150077   122089      121936      121960 119785  85799
#&gt; 4 A2_2021_F1_D0… 134714   109607      109418      109492 107502  79040
#&gt; 5 A2_2021_F1_D1… 151965   122963      122728      122811 120427  89770
#&gt; 6 A2_2021_F1_D1… 158259   125576      125344      125382 123076  92669</code></pre>
</div>
<h3 id="omega_c-2"><code>OMEGA_C = 2</code></h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/rm.html'>rm</a></span><span class='op'>(</span><span class='va'>dadas</span>, <span class='va'>mergers</span>, <span class='va'>seqtab</span>, <span class='va'>seqtab.nochim</span>, <span class='va'>track</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Sample inference w/ <code>dada()</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dadas</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://purrr.tidyverse.org/reference/map2.html'>map2</a></span><span class='op'>(</span>
  <span class='va'>path_filt</span>,
  <span class='va'>errs</span>,
  <span class='va'>dada</span>,
  multithread <span class='op'>=</span> <span class='cn'>TRUE</span>, pool <span class='op'>=</span> <span class='st'>"pseudo"</span>,
  PSEUDO_PREVALENCE <span class='op'>=</span> <span class='fl'>4</span>,
  OMEGA_C <span class='op'>=</span> <span class='fl'>2</span>
<span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Sample 1 - 121742 reads in 25575 unique sequences.
#&gt; Sample 2 - 128950 reads in 26403 unique sequences.
#&gt; Sample 3 - 122089 reads in 23363 unique sequences.
#&gt; Sample 4 - 109607 reads in 24030 unique sequences.
#&gt; Sample 5 - 122963 reads in 26667 unique sequences.
#&gt; Sample 6 - 125576 reads in 27192 unique sequences.
#&gt; Sample 7 - 111264 reads in 24130 unique sequences.
#&gt; Sample 8 - 174377 reads in 35475 unique sequences.
#&gt; Sample 9 - 117765 reads in 26204 unique sequences.
#&gt; Sample 10 - 134307 reads in 28686 unique sequences.
#&gt; Sample 11 - 148672 reads in 30167 unique sequences.
#&gt; Sample 12 - 110742 reads in 25222 unique sequences.
#&gt; Sample 13 - 144767 reads in 27835 unique sequences.
#&gt; Sample 14 - 136169 reads in 24914 unique sequences.
#&gt; Sample 15 - 110253 reads in 22058 unique sequences.
#&gt; Sample 16 - 136410 reads in 24539 unique sequences.
#&gt; Sample 17 - 117137 reads in 22606 unique sequences.
#&gt; Sample 18 - 65569 reads in 17913 unique sequences.
#&gt; Sample 19 - 102670 reads in 20851 unique sequences.
#&gt; Sample 20 - 112281 reads in 24814 unique sequences.
#&gt; Sample 21 - 138049 reads in 27394 unique sequences.
#&gt; Sample 22 - 157679 reads in 28636 unique sequences.
#&gt; Sample 23 - 113866 reads in 22445 unique sequences.
#&gt; Sample 24 - 133996 reads in 26824 unique sequences.
#&gt; Sample 25 - 165095 reads in 30527 unique sequences.
#&gt; Sample 26 - 123661 reads in 25576 unique sequences.
#&gt; Sample 27 - 145073 reads in 29058 unique sequences.
#&gt; Sample 28 - 108056 reads in 23024 unique sequences.
#&gt; Sample 29 - 123552 reads in 26434 unique sequences.
#&gt; Sample 30 - 110229 reads in 25884 unique sequences.
#&gt; Sample 31 - 107530 reads in 22463 unique sequences.
#&gt; Sample 32 - 112419 reads in 25782 unique sequences.
#&gt; Sample 33 - 125104 reads in 26440 unique sequences.
#&gt; Sample 34 - 197929 reads in 35784 unique sequences.
#&gt; Sample 35 - 157754 reads in 30631 unique sequences.
#&gt; Sample 36 - 126564 reads in 28762 unique sequences.
#&gt; Sample 37 - 133185 reads in 27801 unique sequences.
#&gt; Sample 38 - 155321 reads in 31368 unique sequences.
#&gt; Sample 39 - 105466 reads in 23726 unique sequences.
#&gt; Sample 40 - 115309 reads in 25816 unique sequences.
#&gt; Sample 41 - 121892 reads in 25785 unique sequences.
#&gt; Sample 42 - 98848 reads in 22943 unique sequences.
#&gt; Sample 43 - 99851 reads in 24810 unique sequences.
#&gt; Sample 44 - 162855 reads in 31337 unique sequences.
#&gt; Sample 45 - 104604 reads in 22669 unique sequences.
#&gt; Sample 46 - 126851 reads in 27724 unique sequences.
#&gt; Sample 47 - 108314 reads in 23521 unique sequences.
#&gt; Sample 48 - 130225 reads in 28055 unique sequences.
#&gt; Sample 49 - 149520 reads in 27458 unique sequences.
#&gt; Sample 50 - 162053 reads in 31445 unique sequences.
#&gt; Sample 51 - 119936 reads in 24325 unique sequences.
#&gt; Sample 52 - 116195 reads in 24568 unique sequences.
#&gt; Sample 53 - 110691 reads in 24432 unique sequences.
#&gt; Sample 54 - 112178 reads in 27041 unique sequences.
#&gt; Sample 55 - 167359 reads in 36355 unique sequences.
#&gt; Sample 56 - 95833 reads in 22172 unique sequences.
#&gt; Sample 57 - 111376 reads in 27055 unique sequences.
#&gt; Sample 58 - 106360 reads in 23371 unique sequences.
#&gt; Sample 59 - 126877 reads in 32229 unique sequences.
#&gt; Sample 60 - 117907 reads in 25255 unique sequences.
#&gt; Sample 61 - 111267 reads in 25950 unique sequences.
#&gt; Sample 62 - 118149 reads in 25339 unique sequences.
#&gt; Sample 63 - 161806 reads in 30974 unique sequences.
#&gt; Sample 64 - 138744 reads in 27896 unique sequences.
#&gt; Sample 65 - 98773 reads in 23452 unique sequences.
#&gt; Sample 66 - 176975 reads in 32963 unique sequences.
#&gt; Sample 67 - 113222 reads in 24923 unique sequences.
#&gt; Sample 68 - 134841 reads in 30698 unique sequences.
#&gt; Sample 69 - 149102 reads in 32356 unique sequences.
#&gt; Sample 70 - 115392 reads in 26864 unique sequences.
#&gt; Sample 71 - 120571 reads in 24846 unique sequences.
#&gt; Sample 72 - 108729 reads in 23280 unique sequences.
#&gt; Sample 73 - 108661 reads in 22716 unique sequences.
#&gt; Sample 74 - 148569 reads in 30846 unique sequences.
#&gt; Sample 75 - 115256 reads in 25053 unique sequences.
#&gt; Sample 76 - 104048 reads in 24693 unique sequences.
#&gt; Sample 77 - 116114 reads in 25206 unique sequences.
#&gt; Sample 78 - 108593 reads in 21090 unique sequences.
#&gt; Sample 79 - 167701 reads in 34161 unique sequences.
#&gt; Sample 80 - 170878 reads in 37831 unique sequences.
#&gt; Sample 81 - 132214 reads in 30000 unique sequences.
#&gt; Sample 82 - 157762 reads in 30346 unique sequences.
#&gt; Sample 83 - 154395 reads in 31945 unique sequences.
#&gt; Sample 84 - 113625 reads in 25282 unique sequences.
#&gt; Sample 85 - 118923 reads in 25640 unique sequences.
#&gt; Sample 86 - 131569 reads in 30861 unique sequences.
#&gt; Sample 87 - 113192 reads in 26263 unique sequences.
#&gt; Sample 88 - 105216 reads in 23463 unique sequences.
#&gt; Sample 89 - 113874 reads in 25122 unique sequences.
#&gt; Sample 90 - 123707 reads in 27819 unique sequences.
#&gt; Sample 91 - 94532 reads in 27867 unique sequences.
#&gt; Sample 92 - 96498 reads in 26033 unique sequences.
#&gt; Sample 93 - 111858 reads in 23220 unique sequences.
#&gt; Sample 94 - 121948 reads in 24379 unique sequences.
#&gt; Sample 95 - 49178 reads in 16475 unique sequences.
#&gt; Sample 96 - 140587 reads in 26137 unique sequences.
#&gt; 
#&gt;    selfConsist step 2Sample 1 - 121742 reads in 21779 unique sequences.
#&gt; Sample 2 - 128950 reads in 23946 unique sequences.
#&gt; Sample 3 - 122089 reads in 24252 unique sequences.
#&gt; Sample 4 - 109607 reads in 18606 unique sequences.
#&gt; Sample 5 - 122963 reads in 26318 unique sequences.
#&gt; Sample 6 - 125576 reads in 22883 unique sequences.
#&gt; Sample 7 - 111264 reads in 19398 unique sequences.
#&gt; Sample 8 - 174377 reads in 30058 unique sequences.
#&gt; Sample 9 - 117765 reads in 24831 unique sequences.
#&gt; Sample 10 - 134307 reads in 28794 unique sequences.
#&gt; Sample 11 - 148672 reads in 28211 unique sequences.
#&gt; Sample 12 - 110742 reads in 22938 unique sequences.
#&gt; Sample 13 - 144767 reads in 27934 unique sequences.
#&gt; Sample 14 - 136169 reads in 27278 unique sequences.
#&gt; Sample 15 - 110253 reads in 20302 unique sequences.
#&gt; Sample 16 - 136410 reads in 25638 unique sequences.
#&gt; Sample 17 - 117137 reads in 22026 unique sequences.
#&gt; Sample 18 - 65569 reads in 13221 unique sequences.
#&gt; Sample 19 - 102670 reads in 19752 unique sequences.
#&gt; Sample 20 - 112281 reads in 24676 unique sequences.
#&gt; Sample 21 - 138049 reads in 28555 unique sequences.
#&gt; Sample 22 - 157679 reads in 29815 unique sequences.
#&gt; Sample 23 - 113866 reads in 21867 unique sequences.
#&gt; Sample 24 - 133996 reads in 24254 unique sequences.
#&gt; Sample 25 - 165095 reads in 31313 unique sequences.
#&gt; Sample 26 - 123661 reads in 25482 unique sequences.
#&gt; Sample 27 - 145073 reads in 26033 unique sequences.
#&gt; Sample 28 - 108056 reads in 20428 unique sequences.
#&gt; Sample 29 - 123552 reads in 26723 unique sequences.
#&gt; Sample 30 - 110229 reads in 20822 unique sequences.
#&gt; Sample 31 - 107530 reads in 21801 unique sequences.
#&gt; Sample 32 - 112419 reads in 21824 unique sequences.
#&gt; Sample 33 - 125104 reads in 25992 unique sequences.
#&gt; Sample 34 - 197929 reads in 44233 unique sequences.
#&gt; Sample 35 - 157754 reads in 32793 unique sequences.
#&gt; Sample 36 - 126564 reads in 24502 unique sequences.
#&gt; Sample 37 - 133185 reads in 26414 unique sequences.
#&gt; Sample 38 - 155321 reads in 34382 unique sequences.
#&gt; Sample 39 - 105466 reads in 20123 unique sequences.
#&gt; Sample 40 - 115309 reads in 24262 unique sequences.
#&gt; Sample 41 - 121892 reads in 26507 unique sequences.
#&gt; Sample 42 - 98848 reads in 21389 unique sequences.
#&gt; Sample 43 - 99851 reads in 18974 unique sequences.
#&gt; Sample 44 - 162855 reads in 30391 unique sequences.
#&gt; Sample 45 - 104604 reads in 18900 unique sequences.
#&gt; Sample 46 - 126851 reads in 26866 unique sequences.
#&gt; Sample 47 - 108314 reads in 22909 unique sequences.
#&gt; Sample 48 - 130225 reads in 22975 unique sequences.
#&gt; Sample 49 - 149520 reads in 33157 unique sequences.
#&gt; Sample 50 - 162053 reads in 33715 unique sequences.
#&gt; Sample 51 - 119936 reads in 23859 unique sequences.
#&gt; Sample 52 - 116195 reads in 24250 unique sequences.
#&gt; Sample 53 - 110691 reads in 23385 unique sequences.
#&gt; Sample 54 - 112178 reads in 32963 unique sequences.
#&gt; Sample 55 - 167359 reads in 30380 unique sequences.
#&gt; Sample 56 - 95833 reads in 17799 unique sequences.
#&gt; Sample 57 - 111376 reads in 20483 unique sequences.
#&gt; Sample 58 - 106360 reads in 20752 unique sequences.
#&gt; Sample 59 - 126877 reads in 22471 unique sequences.
#&gt; Sample 60 - 117907 reads in 25967 unique sequences.
#&gt; Sample 61 - 111267 reads in 21427 unique sequences.
#&gt; Sample 62 - 118149 reads in 22154 unique sequences.
#&gt; Sample 63 - 161806 reads in 31608 unique sequences.
#&gt; Sample 64 - 138744 reads in 27484 unique sequences.
#&gt; Sample 65 - 98773 reads in 18508 unique sequences.
#&gt; Sample 66 - 176975 reads in 37865 unique sequences.
#&gt; Sample 67 - 113222 reads in 21674 unique sequences.
#&gt; Sample 68 - 134841 reads in 25149 unique sequences.
#&gt; Sample 69 - 149102 reads in 27194 unique sequences.
#&gt; Sample 70 - 115392 reads in 21553 unique sequences.
#&gt; Sample 71 - 120571 reads in 21898 unique sequences.
#&gt; Sample 72 - 108729 reads in 20542 unique sequences.
#&gt; Sample 73 - 108661 reads in 21550 unique sequences.
#&gt; Sample 74 - 148569 reads in 28137 unique sequences.
#&gt; Sample 75 - 115256 reads in 21853 unique sequences.
#&gt; Sample 76 - 104048 reads in 20642 unique sequences.
#&gt; Sample 77 - 116114 reads in 20887 unique sequences.
#&gt; Sample 78 - 108593 reads in 21800 unique sequences.
#&gt; Sample 79 - 167701 reads in 35267 unique sequences.
#&gt; Sample 80 - 170878 reads in 31030 unique sequences.
#&gt; Sample 81 - 132214 reads in 24867 unique sequences.
#&gt; Sample 82 - 157762 reads in 35675 unique sequences.
#&gt; Sample 83 - 154395 reads in 33985 unique sequences.
#&gt; Sample 84 - 113625 reads in 22264 unique sequences.
#&gt; Sample 85 - 118923 reads in 25540 unique sequences.
#&gt; Sample 86 - 131569 reads in 29033 unique sequences.
#&gt; Sample 87 - 113192 reads in 23522 unique sequences.
#&gt; Sample 88 - 105216 reads in 23531 unique sequences.
#&gt; Sample 89 - 113874 reads in 23178 unique sequences.
#&gt; Sample 90 - 123707 reads in 24794 unique sequences.
#&gt; Sample 91 - 94532 reads in 27889 unique sequences.
#&gt; Sample 92 - 96498 reads in 26403 unique sequences.
#&gt; Sample 93 - 111858 reads in 23120 unique sequences.
#&gt; Sample 94 - 121948 reads in 23388 unique sequences.
#&gt; Sample 95 - 49178 reads in 30282 unique sequences.
#&gt; Sample 96 - 140587 reads in 25653 unique sequences.
#&gt; 
#&gt;    selfConsist step 2</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>dadas</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"dadas-2.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Check the results:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code>#&gt; dada-class: object describing DADA2 denoising results
#&gt; 456 sequence variants were inferred from 25575 
#&gt; input unique sequences.
#&gt; Key parameters: OMEGA_A = 1e-40, OMEGA_C = 2, BAND_SIZE = 16</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getSequences.html'>getSequences</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>, <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;TGAGGAATATTGGTCAATGGGCGAGAGCCTGAACCAGCCAAGTAGCGTGAAGGATGAAGGCTCTATGGGTCGTAAACTTCTTTTATATGGGAATAAAGTTTTCCACGTGTGGAATTTTGTATGTACCATATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGATTGTTAAGTCAGTTGTGAAAGTTTGCGGCTCA&quot;
#&gt; [2] &quot;TGAGGAATATTGGTCAATGGGCGCAGGCCTGAACCAGCCAAGTAGCGTGAAGGATGACTGCCCTATGGGTTGTAAACTTCTTTTATATGGGAATAAAGTTTTCCACGTGTGGAATTTTGTATGTACCATATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGACAGTTAAGTCAGTTGTGAAAGTTTGCGGCTCA&quot;</code></pre>
</div>
<p>Merge paired reads,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>mergers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/mergePairs.html'>mergePairs</a></span><span class='op'>(</span>
  <span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, 
  <span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>path_filt</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>mergers</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"mergers-2.Rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Construct sequence table</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>seqtab</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/makeSequenceTable.html'>makeSequenceTable</a></span><span class='op'>(</span><span class='va'>mergers</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>seqtab</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"seqtab-2.Rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1]    96 24402</code></pre>
</div>
<p>Inspect distribution of sequence lengths</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nchar.html'>nchar</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getSequences.html'>getSequences</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; 
#&gt;   259   260   267   269   272   276   284   323   325   327   328 
#&gt;    58    20    29     1     3     1     1     1     2     2     1 
#&gt;   331   335   338   341   343   344   346   348   349   350   351 
#&gt;     1     7     1     3     3     5     1     1     1     5     1 
#&gt;   358   364   370   372   373   377   378   379   380   383   384 
#&gt;     1     1     5     1     1     1     1     1     1     7    10 
#&gt;   386   387   388   389   392   393   398   399   400   401   402 
#&gt;    10     1     1     4    11     1     8     2     1    80  1716 
#&gt;   403   404   405   406   407   408   409   411   413   414   415 
#&gt;   389   372    16    10   100   905     5     1     1     4     1 
#&gt;   419   420   421   422   423   424   425   426   427   428   429 
#&gt;     2    12   125 14250    16     3     7    55  5197   890    10 
#&gt;   432   438   439   441   444   445   446   447   448 
#&gt;     1     1     1     3     3     2     3     3     1</code></pre>
</div>
<p>Remove chimeras (with ‘pooled’ method)</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>seqtab.nochim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/dada2/man/removeBimeraDenovo.html'>removeBimeraDenovo</a></span><span class='op'>(</span><span class='va'>seqtab</span>, method <span class='op'>=</span> <span class='st'>"pooled"</span>,
    multithread <span class='op'>=</span> <span class='cn'>TRUE</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; Identified 22346 bimeras out of 24402 input sequences.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"seqtab-nochim-2.Rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1]   96 2056</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>seqtab</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] 0.7875494</code></pre>
</div>
<p>Track reads through the pipeline,</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>getN</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/dada2/man/getUniques.html'>getUniques</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>track</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>out</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>getN</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>dadas</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>getN</span><span class='op'>)</span>, 
    <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>mergers</span>, <span class='va'>getN</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='va'>seqtab.nochim</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"input"</span>, <span class='st'>"filtered"</span>, <span class='st'>"denoised_r1"</span>, <span class='st'>"denoised_r2"</span>, <span class='st'>"merged"</span>,
    <span class='st'>"nochim"</span><span class='op'>)</span>
<span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>ftb</span><span class='op'>$</span><span class='va'>sample_id</span>
<span class='va'>track</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://tibble.tidyverse.org/reference/as_tibble.html'>as_tibble</a></span><span class='op'>(</span><span class='va'>track</span>, rownames <span class='op'>=</span> <span class='st'>"sample_id"</span><span class='op'>)</span>
<span class='fu'><a href='https://readr.tidyverse.org/reference/write_delim.html'>write_csv</a></span><span class='op'>(</span><span class='va'>track</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>results_path</span>, <span class='st'>"track-2.csv"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>track</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 6 × 7
#&gt;   sample_id       input filtered denoised_r1 denoised_r2 merged nochim
#&gt;   &lt;chr&gt;           &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
#&gt; 1 A2_2021_F1_D0… 150600   121742       82879       88192  66517  47118
#&gt; 2 A2_2021_F1_D0… 160932   128950       87961       91752  69451  51802
#&gt; 3 A2_2021_F1_D0… 150077   122089       85573       85424  66136  46854
#&gt; 4 A2_2021_F1_D0… 134714   109607       73664       80377  59143  43190
#&gt; 5 A2_2021_F1_D1… 151965   122963       83052       84053  63282  46862
#&gt; 6 A2_2021_F1_D1… 158259   125576       83871       89685  66456  49876</code></pre>
</div>
<h2 id="session-info">Session info</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://r-lib.github.io/sessioninfo/reference/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────
#&gt;  setting  value
#&gt;  version  R version 4.1.2 (2021-11-01)
#&gt;  os       Arch Linux
#&gt;  system   x86_64, linux-gnu
#&gt;  ui       X11
#&gt;  language (EN)
#&gt;  collate  en_US.UTF-8
#&gt;  ctype    en_US.UTF-8
#&gt;  tz       America/New_York
#&gt;  date     2022-02-18
#&gt;  pandoc   2.14.1 @ /usr/bin/ (via rmarkdown)
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────
#&gt;  package              * version  date (UTC) lib source
#&gt;  assertthat             0.2.1    2019-03-21 [2] CRAN (R 4.0.0)
#&gt;  backports              1.4.1    2021-12-13 [2] CRAN (R 4.1.2)
#&gt;  Biobase                2.52.0   2021-05-19 [2] Bioconductor
#&gt;  BiocGenerics           0.38.0   2021-05-19 [2] Bioconductor
#&gt;  BiocParallel           1.26.2   2021-08-22 [2] Bioconductor
#&gt;  Biostrings             2.60.2   2021-08-05 [2] Bioconductor
#&gt;  bit                    4.0.4    2020-08-04 [2] CRAN (R 4.0.2)
#&gt;  bit64                  4.0.5    2020-08-30 [2] CRAN (R 4.0.2)
#&gt;  bitops                 1.0-7    2021-04-24 [2] CRAN (R 4.1.0)
#&gt;  blob                   1.2.2    2021-07-23 [2] CRAN (R 4.1.0)
#&gt;  brio                   1.1.3    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  broom                  0.7.11   2022-01-03 [2] CRAN (R 4.1.2)
#&gt;  bslib                  0.3.1    2021-10-06 [2] CRAN (R 4.1.1)
#&gt;  cachem                 1.0.6    2021-08-19 [2] CRAN (R 4.1.1)
#&gt;  callr                  3.7.0    2021-04-20 [2] CRAN (R 4.1.0)
#&gt;  cellranger             1.1.0    2016-07-27 [2] CRAN (R 4.0.0)
#&gt;  cli                    3.2.0    2022-02-14 [2] CRAN (R 4.1.2)
#&gt;  colorspace             2.0-2    2021-08-11 [2] R-Forge (R 4.1.1)
#&gt;  crayon                 1.5.0    2022-02-14 [2] CRAN (R 4.1.2)
#&gt;  dada2                * 1.18.0   2021-09-18 [1] Github (benjjneb/dada2@a20a676)
#&gt;  DBI                    1.1.2    2021-12-20 [2] CRAN (R 4.1.2)
#&gt;  dbplyr                 2.1.1    2021-04-06 [2] CRAN (R 4.0.5)
#&gt;  DECIPHER               2.20.0   2021-05-19 [2] Bioconductor
#&gt;  DelayedArray           0.18.0   2021-05-19 [2] Bioconductor
#&gt;  desc                   1.4.0    2021-09-28 [2] CRAN (R 4.1.1)
#&gt;  devtools               2.4.3    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  digest                 0.6.29   2021-12-01 [2] CRAN (R 4.1.2)
#&gt;  distill                1.3      2021-10-13 [2] CRAN (R 4.1.1)
#&gt;  downlit                0.4.0    2021-10-29 [2] CRAN (R 4.1.2)
#&gt;  dplyr                * 1.0.8    2022-02-08 [2] CRAN (R 4.1.2)
#&gt;  ellipsis               0.3.2    2021-04-29 [2] CRAN (R 4.1.0)
#&gt;  evaluate               0.14     2019-05-28 [2] CRAN (R 4.0.0)
#&gt;  fansi                  1.0.2    2022-01-14 [2] CRAN (R 4.1.2)
#&gt;  farver                 2.1.0    2021-02-28 [2] CRAN (R 4.0.4)
#&gt;  fastmap                1.1.0    2021-01-25 [2] CRAN (R 4.0.4)
#&gt;  forcats              * 0.5.1    2021-01-27 [2] CRAN (R 4.0.4)
#&gt;  fs                   * 1.5.2    2021-12-08 [2] CRAN (R 4.1.2)
#&gt;  generics               0.1.2    2022-01-31 [2] CRAN (R 4.1.2)
#&gt;  GenomeInfoDb           1.28.4   2021-09-05 [2] Bioconductor
#&gt;  GenomeInfoDbData       1.2.6    2021-05-31 [2] Bioconductor
#&gt;  GenomicAlignments      1.28.0   2021-05-19 [2] Bioconductor
#&gt;  GenomicRanges          1.44.0   2021-05-19 [2] Bioconductor
#&gt;  ggplot2              * 3.3.5    2021-06-25 [2] CRAN (R 4.1.0)
#&gt;  glue                   1.6.1    2022-01-22 [2] CRAN (R 4.1.2)
#&gt;  gtable                 0.3.0    2019-03-25 [2] CRAN (R 4.0.0)
#&gt;  haven                  2.4.3    2021-08-04 [2] CRAN (R 4.1.1)
#&gt;  here                 * 1.0.1    2020-12-13 [2] CRAN (R 4.0.5)
#&gt;  highr                  0.9      2021-04-16 [2] CRAN (R 4.1.0)
#&gt;  hms                    1.1.1    2021-09-26 [2] CRAN (R 4.1.1)
#&gt;  htmltools              0.5.2    2021-08-25 [2] CRAN (R 4.1.1)
#&gt;  httr                   1.4.2    2020-07-20 [2] CRAN (R 4.0.2)
#&gt;  hwriter                1.3.2    2014-09-10 [2] CRAN (R 4.0.0)
#&gt;  import                 1.2.0    2020-09-24 [2] CRAN (R 4.0.2)
#&gt;  IRanges                2.26.0   2021-05-19 [2] Bioconductor
#&gt;  jpeg                   0.1-9    2021-07-24 [2] CRAN (R 4.1.0)
#&gt;  jquerylib              0.1.4    2021-04-26 [2] CRAN (R 4.1.0)
#&gt;  jsonlite               1.7.3    2022-01-17 [2] CRAN (R 4.1.2)
#&gt;  knitr                  1.37     2021-12-16 [2] CRAN (R 4.1.2)
#&gt;  labeling               0.4.2    2020-10-20 [2] CRAN (R 4.0.3)
#&gt;  lattice                0.20-45  2021-09-22 [3] CRAN (R 4.1.2)
#&gt;  latticeExtra           0.6-29   2019-12-19 [2] CRAN (R 4.0.0)
#&gt;  lifecycle              1.0.1    2021-09-24 [2] CRAN (R 4.1.1)
#&gt;  lubridate              1.8.0    2021-10-07 [2] CRAN (R 4.1.1)
#&gt;  magrittr               2.0.2    2022-01-26 [2] CRAN (R 4.1.2)
#&gt;  Matrix                 1.3-4    2021-06-01 [3] CRAN (R 4.1.2)
#&gt;  MatrixGenerics         1.4.3    2021-08-26 [2] Bioconductor
#&gt;  matrixStats            0.61.0   2021-09-17 [2] CRAN (R 4.1.1)
#&gt;  memoise                2.0.1    2021-11-26 [2] CRAN (R 4.1.2)
#&gt;  modelr                 0.1.8    2020-05-19 [2] CRAN (R 4.0.0)
#&gt;  munsell                0.5.0    2018-06-12 [2] CRAN (R 4.0.0)
#&gt;  nvimcom              * 0.9-102  2021-11-12 [2] local
#&gt;  pillar                 1.7.0    2022-02-01 [2] CRAN (R 4.1.2)
#&gt;  pkgbuild               1.3.1    2021-12-20 [2] CRAN (R 4.1.2)
#&gt;  pkgconfig              2.0.3    2019-09-22 [2] CRAN (R 4.0.0)
#&gt;  pkgload                1.2.4    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  plyr                   1.8.6    2020-03-03 [2] CRAN (R 4.0.0)
#&gt;  png                    0.1-7    2013-12-03 [2] CRAN (R 4.0.0)
#&gt;  prettyunits            1.1.1    2020-01-24 [2] CRAN (R 4.0.0)
#&gt;  processx               3.5.2    2021-04-30 [2] CRAN (R 4.1.0)
#&gt;  ps                     1.6.0    2021-02-28 [2] CRAN (R 4.0.4)
#&gt;  purrr                * 0.3.4    2020-04-17 [2] CRAN (R 4.0.0)
#&gt;  R6                     2.5.1    2021-08-19 [2] CRAN (R 4.1.1)
#&gt;  ragg                   1.2.1    2021-12-06 [2] CRAN (R 4.1.2)
#&gt;  RColorBrewer           1.1-2    2014-12-07 [2] CRAN (R 4.0.0)
#&gt;  Rcpp                 * 1.0.8    2022-01-13 [2] CRAN (R 4.1.2)
#&gt;  RcppParallel           5.1.5    2022-01-05 [2] CRAN (R 4.1.2)
#&gt;  RCurl                  1.98-1.5 2021-09-17 [2] CRAN (R 4.1.1)
#&gt;  readr                * 2.1.1    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  readxl                 1.3.1    2019-03-13 [2] CRAN (R 4.0.0)
#&gt;  remotes                2.4.2    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  reprex                 2.0.1    2021-08-05 [2] CRAN (R 4.1.1)
#&gt;  reshape2               1.4.4    2020-04-09 [2] CRAN (R 4.0.0)
#&gt;  rlang                  1.0.1    2022-02-03 [2] CRAN (R 4.1.2)
#&gt;  rmarkdown            * 2.11     2021-09-14 [2] CRAN (R 4.1.1)
#&gt;  rprojroot              2.0.2    2020-11-15 [2] CRAN (R 4.0.3)
#&gt;  Rsamtools              2.8.0    2021-05-19 [2] Bioconductor
#&gt;  RSQLite                2.2.9    2021-12-06 [2] CRAN (R 4.1.2)
#&gt;  rstudioapi             0.13     2020-11-12 [2] CRAN (R 4.0.3)
#&gt;  rvest                  1.0.2    2021-10-16 [2] CRAN (R 4.1.1)
#&gt;  S4Vectors              0.30.2   2021-10-03 [2] Bioconductor
#&gt;  sass                   0.4.0    2021-05-12 [2] CRAN (R 4.1.0)
#&gt;  scales                 1.1.1    2020-05-11 [2] CRAN (R 4.0.0)
#&gt;  sessioninfo            1.2.2    2021-12-06 [2] CRAN (R 4.1.2)
#&gt;  ShortRead              1.50.0   2021-05-19 [2] Bioconductor
#&gt;  stringi                1.7.6    2021-11-29 [2] CRAN (R 4.1.2)
#&gt;  stringr              * 1.4.0    2019-02-10 [2] CRAN (R 4.0.0)
#&gt;  SummarizedExperiment   1.22.0   2021-05-19 [2] Bioconductor
#&gt;  systemfonts            1.0.3    2021-10-13 [2] CRAN (R 4.1.1)
#&gt;  testthat               3.1.2    2022-01-20 [2] CRAN (R 4.1.2)
#&gt;  textshaping            0.3.6    2021-10-13 [2] CRAN (R 4.1.1)
#&gt;  tibble               * 3.1.6    2021-11-07 [2] CRAN (R 4.1.2)
#&gt;  tidyr                * 1.2.0    2022-02-01 [2] CRAN (R 4.1.2)
#&gt;  tidyselect             1.1.1    2021-04-30 [2] CRAN (R 4.1.0)
#&gt;  tidyverse            * 1.3.1    2021-04-15 [2] CRAN (R 4.1.0)
#&gt;  tzdb                   0.2.0    2021-10-27 [2] CRAN (R 4.1.2)
#&gt;  usethis                2.1.5    2021-12-09 [2] CRAN (R 4.1.2)
#&gt;  utf8                   1.2.2    2021-07-24 [2] CRAN (R 4.1.0)
#&gt;  vctrs                  0.3.8    2021-04-29 [2] CRAN (R 4.1.0)
#&gt;  vroom                  1.5.7    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  withr                  2.4.3    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  xfun                   0.29     2021-12-14 [2] CRAN (R 4.1.2)
#&gt;  xml2                   1.3.3    2021-11-30 [2] CRAN (R 4.1.2)
#&gt;  XVector                0.32.0   2021-05-19 [2] Bioconductor
#&gt;  yaml                   2.2.2    2022-01-25 [2] CRAN (R 4.1.2)
#&gt;  zlibbioc               1.38.0   2021-05-19 [2] Bioconductor
#&gt; 
#&gt;  [1] /home/michael/research/vivo-vitro/lib/R/library-for-dada2-1.18.0
#&gt;  [2] /home/michael/.local/lib/R/library
#&gt;  [3] /usr/lib/R/library
#&gt; 
#&gt; ──────────────────────────────────────────────────────────────────────────</code></pre>
</div>
<div class="sourceCode" id="cb42"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
